// bdsx/bds/actor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Actor = exports.ActorType = exports.ActorRuntimeID = exports.DimensionId = exports.ActorUniqueID = void 0;
const bin_1 = require("bdsx/bin");
const common_1 = require("bdsx/common");
const core_1 = require("bdsx/core");
const makefunc_1 = require("bdsx/makefunc");
const nativeclass_1 = require("bdsx/nativeclass");
const nativetype_1 = require("bdsx/nativetype");
const pointer_1 = require("bdsx/pointer");
const attribute_1 = require("./attribute");
exports.ActorUniqueID = nativetype_1.bin64_t.extends();
var DimensionId;
(function (DimensionId) {
    DimensionId[DimensionId["Overworld"] = 0] = "Overworld";
    DimensionId[DimensionId["Nether"] = 1] = "Nether";
    DimensionId[DimensionId["TheEnd"] = 2] = "TheEnd";
})(DimensionId = exports.DimensionId || (exports.DimensionId = {}));
class ActorRuntimeID extends core_1.VoidPointer {
}
exports.ActorRuntimeID = ActorRuntimeID;
var ActorType;
(function (ActorType) {
    ActorType[ActorType["Player"] = 319] = "Player";
})(ActorType = exports.ActorType || (exports.ActorType = {}));
class Actor extends nativeclass_1.NativeClass {
    _getName() {
        common_1.abstract();
    }
    _addTag(tag) {
        common_1.abstract();
    }
    _hasTag(tag) {
        common_1.abstract();
    }
    _sendNetworkPacket(packet) {
        common_1.abstract();
    }
    _sendAttributePacket(id, value, attr) {
        common_1.abstract();
    }
    sendPacket(packet) {
        if (!this.isPlayer())
            throw Error("this is not ServerPlayer");
        this._sendNetworkPacket(packet);
    }
    //     static fromPointer(ptr:StaticPointer):Actor;
    //     static fromUniqueId(_64bit_low:number, _64bit_high:number):Actor|null;
    _getDimensionId(out) {
        common_1.abstract();
    }
    getDimension() {
        const out = new Int32Array(1);
        this._getDimensionId(out);
        return out[0];
    }
    /**
     * @deprecated use actor.identifier
     */
    getIdentifier() {
        return this.identifier;
    }
    isPlayer() {
        common_1.abstract();
    }
    getName() {
        return this._getName().value;
    }
    getNetworkIdentifier() {
        throw Error(`this is not player`);
    }
    getPosition() {
        common_1.abstract();
    }
    getRegion() {
        common_1.abstract();
    }
    getUniqueIdLow() {
        return this.getUniqueIdPointer().getInt32(0);
    }
    getUniqueIdHigh() {
        return this.getUniqueIdPointer().getInt32(4);
    }
    getUniqueIdBin() {
        return this.getUniqueIdPointer().getBin64();
    }
    /**
     * it returns address of the unique id field
     */
    getUniqueIdPointer() {
        common_1.abstract();
    }
    getTypeId() {
        common_1.abstract();
    }
    getAttribute(id) {
        const attr = this.attributes.getMutableInstance(id);
        if (attr === null)
            return 0;
        return attr.currentValue;
    }
    setAttribute(id, value) {
        if (id < 1)
            return;
        if (id > 15)
            return;
        const attr = this.attributes.getMutableInstance(id);
        if (attr === null)
            throw Error(`${this.identifier} has not ${attribute_1.AttributeId[id] || 'Attribute' + id}`);
        attr.currentValue = value;
        if (this.isPlayer()) {
            this._sendAttributePacket(id, value, attr);
        }
    }
    /**
     * @deprecated use actor.runtimeId
     */
    getRuntimeId() {
        return this.runtimeId.add();
    }
    /**
     * @deprecated Need more implement
     */
    getEntity() {
        let entity = this.entity;
        if (entity)
            return entity;
        entity = {
            __unique_id__: {
                "64bit_low": this.getUniqueIdLow(),
                "64bit_high": this.getUniqueIdHigh()
            },
            __identifier__: this.identifier,
            __type__: (this.getTypeId() & 0xff) === 0x40 ? 'item_entity' : 'entity',
            id: 0, // bool ScriptApi::WORKAROUNDS::helpRegisterActor(entt::Registry<unsigned int>* registry? ,Actor* actor,unsigned int* id_out);
        };
        return this.entity = entity;
    }
    addTag(tag) {
        const _tag = new pointer_1.CxxStringWrapper(true);
        _tag.construct();
        _tag.value = tag;
        const ret = this._addTag(_tag);
        _tag.destruct();
        return ret;
    }
    hasTag(tag) {
        const _tag = new pointer_1.CxxStringWrapper(true);
        _tag.construct();
        _tag.value = tag;
        const ret = this._hasTag(_tag);
        _tag.destruct();
        return ret;
    }
    // float NativeActor::getAttribute(int attribute) noexcept
    // {
    //     if (attribute < 1) return 0;
    //     if ((uint)attribute > countof(attribNames)) return 0;
    //     AttributeInstance* attr = ptr()->getAttribute((AttributeId)attribute);
    //     if (!attr) return 0;
    //     return attr->currentValue();
    // }
    // kr::JsValue NativeActor::fromPointer(StaticPointer* ptr) throws(JsException)
    // {
    //     if (ptr == nullptr) throw JsException(u"1st argument must be *Pointer");
    //     Actor* actor = (Actor*)ptr->getAddressRaw();
    //     return fromRaw(actor);
    // }
    static fromUniqueIdBin(bin) {
        common_1.abstract();
    }
    static fromUniqueId(lowbits, highbits) {
        return Actor.fromUniqueIdBin(bin_1.bin.make64(lowbits, highbits));
    }
    static fromEntity(entity) {
        const u = entity.__unique_id__;
        return Actor.fromUniqueId(u["64bit_low"], u["64bit_high"]);
    }
    static [makefunc_1.makefunc.np2js](ptr) {
        return Actor._singletoning(ptr);
    }
    static all() {
        common_1.abstract();
    }
    static _singletoning(ptr) {
        common_1.abstract();
    }
}
exports.Actor = Actor;
