// bdsx/bds/block.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlockSource = exports.Block = exports.BlockLegacy = void 0;
const nativeclass_1 = require("bdsx/nativeclass");
const common_1 = require("bdsx/common");
class BlockLegacy extends nativeclass_1.NativeClass {
    _getCommandName() {
        common_1.abstract();
    }
    getCommandName() {
        return this._getCommandName().value;
    }
    getCreativeCategory() {
        common_1.abstract();
    }
    /**
     * Will not affect actual destroy time but will affect the speed of cracks
     */
    setDestroyTime(time) {
        common_1.abstract();
    }
    getDestroyTime() {
        return this.getFloat32(0x12C);
    }
}
exports.BlockLegacy = BlockLegacy;
class Block extends nativeclass_1.NativeClass {
    _getName() {
        common_1.abstract();
    }
    getName() {
        return this._getName().str;
    }
}
exports.Block = Block;
class BlockSource extends nativeclass_1.NativeClass {
    getBlock(blockPos) {
        common_1.abstract();
    }
}
exports.BlockSource = BlockSource;
