// bdsx/bds/commandorigin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerCommandOrigin = exports.ScriptCommandOrigin = exports.PlayerCommandOrigin = exports.CommandOrigin = void 0;
const common_1 = require("bdsx/common");
const mce_1 = require("bdsx/mce");
const nativeclass_1 = require("bdsx/nativeclass");
class CommandOrigin extends nativeclass_1.NativeClass {
    construct(vftable, level) {
        if (vftable !== undefined && level !== undefined) {
            this.vftable = vftable;
            this.level = level;
            this.uuid = mce_1.mce.UUID.generate();
        }
        else {
            super.construct();
        }
    }
    isServerCommandOrigin() {
        common_1.abstract();
    }
    /**
     * @deprecated use cmdorigin.destruct();
     */
    destructor() {
        common_1.abstract();
    }
    getRequestId() {
        common_1.abstract();
    }
    getName() {
        common_1.abstract();
    }
    getBlockPosition() {
        common_1.abstract();
    }
    getWorldPosition() {
        common_1.abstract();
    }
    getLevel() {
        common_1.abstract();
    }
    getDimension() {
        common_1.abstract();
    }
    getEntity() {
        common_1.abstract();
    }
}
exports.CommandOrigin = CommandOrigin;
class PlayerCommandOrigin extends CommandOrigin {
}
exports.PlayerCommandOrigin = PlayerCommandOrigin;
class ScriptCommandOrigin extends PlayerCommandOrigin {
}
exports.ScriptCommandOrigin = ScriptCommandOrigin;
class ServerCommandOrigin extends CommandOrigin {
}
exports.ServerCommandOrigin = ServerCommandOrigin;
