// bdsx/bds/command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommandOutputSender = exports.CommandRegistry = exports.MinecraftCommands = exports.CommandContext = exports.MCRESULT = exports.CommandFlag = exports.CommandPermissionLevel = void 0;
const common_1 = require("../common");
const nativeclass_1 = require("../nativeclass");
const nativetype_1 = require("../nativetype");
const pointer_1 = require("../pointer");
const commandorigin_1 = require("./commandorigin");
var CommandPermissionLevel;
(function (CommandPermissionLevel) {
    CommandPermissionLevel[CommandPermissionLevel["Normal"] = 0] = "Normal";
    CommandPermissionLevel[CommandPermissionLevel["Operator"] = 1] = "Operator";
    CommandPermissionLevel[CommandPermissionLevel["Host"] = 2] = "Host";
    CommandPermissionLevel[CommandPermissionLevel["Automation"] = 3] = "Automation";
    CommandPermissionLevel[CommandPermissionLevel["Admin"] = 4] = "Admin";
})(CommandPermissionLevel = exports.CommandPermissionLevel || (exports.CommandPermissionLevel = {}));
var CommandFlag;
(function (CommandFlag) {
    CommandFlag[CommandFlag["None"] = 0] = "None";
})(CommandFlag = exports.CommandFlag || (exports.CommandFlag = {}));
class MCRESULT extends nativeclass_1.NativeClass {
}
exports.MCRESULT = MCRESULT;
MCRESULT.define({
    result: nativetype_1.uint32_t
});
class CommandContext extends nativeclass_1.NativeClass {
}
exports.CommandContext = CommandContext;
CommandContext.abstract({
    command: nativetype_1.CxxString,
    origin: commandorigin_1.ServerCommandOrigin.ref(),
}, 0x30); // dependency problem, SharedPtr.make(CommandContext) needs the structure defination.
class MinecraftCommands extends nativeclass_1.NativeClass {
    executeCommand(ctx, b) {
        common_1.abstract();
    }
}
exports.MinecraftCommands = MinecraftCommands;
class CommandRegistry extends nativeclass_1.NativeClass {
    _registerCommand(command, description, level, flag1, flag2) {
        common_1.abstract();
    }
    registerCommand(command, description, level, flag1, flag2) {
        const commandstr = new pointer_1.CxxStringWrapper(true);
        commandstr.construct();
        commandstr.value = command;
        this._registerCommand(commandstr, description, level, flag1, flag2);
        commandstr.destruct();
    }
    registerOverloadInternal(signature, overload) {
        common_1.abstract();
    }
    _findCommand(command) {
        common_1.abstract();
    }
    findCommand(command) {
        const commandstr = new pointer_1.CxxStringWrapper(true);
        commandstr.construct();
        commandstr.value = command;
        const sig = this._findCommand(commandstr);
        commandstr.destruct();
        return sig;
    }
}
exports.CommandRegistry = CommandRegistry;
(function (CommandRegistry) {
    class Signature extends nativeclass_1.NativeClass {
    }
    CommandRegistry.Signature = Signature;
    class Overload extends nativeclass_1.NativeClass {
    }
    CommandRegistry.Overload = Overload;
})(CommandRegistry = exports.CommandRegistry || (exports.CommandRegistry = {}));
class CommandOutputSender extends nativeclass_1.NativeClass {
}
exports.CommandOutputSender = CommandOutputSender;
