// bdsx/bds/inventory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlayerInventory = exports.ItemStack = exports.Item = exports.CreativeItemCategory = exports.ContainerId = void 0;
const common_1 = require("bdsx/common");
const nativeclass_1 = require("bdsx/nativeclass");
const pointer_1 = require("bdsx/pointer");
var ContainerId;
(function (ContainerId) {
    ContainerId[ContainerId["Inventory"] = 0] = "Inventory";
    ContainerId[ContainerId["First"] = 1] = "First";
    ContainerId[ContainerId["Last"] = 100] = "Last";
    ContainerId[ContainerId["Offhand"] = 119] = "Offhand";
    ContainerId[ContainerId["Armor"] = 120] = "Armor";
    ContainerId[ContainerId["Hotbar"] = 122] = "Hotbar";
    ContainerId[ContainerId["FixedInventory"] = 123] = "FixedInventory";
    ContainerId[ContainerId["UI"] = 124] = "UI";
})(ContainerId = exports.ContainerId || (exports.ContainerId = {}));
var CreativeItemCategory;
(function (CreativeItemCategory) {
    CreativeItemCategory[CreativeItemCategory["Construction"] = 1] = "Construction";
    CreativeItemCategory[CreativeItemCategory["Nature"] = 2] = "Nature";
    CreativeItemCategory[CreativeItemCategory["Items"] = 4] = "Items";
    CreativeItemCategory[CreativeItemCategory["Uncategorized"] = 5] = "Uncategorized";
})(CreativeItemCategory = exports.CreativeItemCategory || (exports.CreativeItemCategory = {}));
class Item extends nativeclass_1.NativeClass {
    _getCommandName() {
        common_1.abstract();
    }
    allowOffhand() {
        common_1.abstract();
    }
    getCommandName() {
        return this._getCommandName().value;
    }
    getCreativeCategory() {
        common_1.abstract();
    }
    isDamageable() {
        common_1.abstract();
    }
    isFood() {
        common_1.abstract();
    }
    /**
     * Will not affect client but allows /replaceitem
     */
    setAllowOffhand(value) {
        common_1.abstract();
    }
}
exports.Item = Item;
class ItemStack extends nativeclass_1.NativeClass {
    _getId() {
        common_1.abstract();
    }
    _getItem() {
        common_1.abstract();
    }
    _setCustomName(name) {
        common_1.abstract();
    }
    isBlock() {
        common_1.abstract();
    }
    isNull() {
        common_1.abstract();
    }
    getAmount() {
        return this.amount;
    }
    getId() {
        const id = this._getId();
        if (id > 32767) {
            return 255 + (65536 - id); // TODO: Fix this when there is a wrapper for Short
        }
        return id;
    }
    getItem() {
        if (this.isNull()) {
            return null;
        }
        return this._getItem();
    }
    getName() {
        var _a, _b;
        return (_b = "minecraft:" + ((_a = this.getItem()) === null || _a === void 0 ? void 0 : _a.getCommandName())) !== null && _b !== void 0 ? _b : "air";
    }
    hasCustomName() {
        common_1.abstract();
    }
    setAmount(amount) {
        this.amount = amount;
    }
    setCustomName(name) {
        const _name = new pointer_1.CxxStringWrapper(true);
        _name.construct();
        _name.value = name;
        this._setCustomName(_name);
        _name.destruct();
    }
}
exports.ItemStack = ItemStack;
class PlayerInventory extends nativeclass_1.NativeClass {
    getItem(slot, containerId) {
        common_1.abstract();
    }
}
exports.PlayerInventory = PlayerInventory;
