// bdsx/bds/networkidentifier.js
"use strict";
var NetworkIdentifier_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.networkHandler = exports.NetworkIdentifier = exports.ServerNetworkHandler = exports.NetworkHandler = void 0;
const tslib_1 = require("tslib");
const assembler_1 = require("bdsx/assembler");
const common_1 = require("bdsx/common");
const dll_1 = require("bdsx/dll");
const hashset_1 = require("bdsx/hashset");
const makefunc_1 = require("bdsx/makefunc");
const nativeclass_1 = require("bdsx/nativeclass");
const nativetype_1 = require("bdsx/nativetype");
const pointer_1 = require("bdsx/pointer");
const source_map_support_1 = require("bdsx/source-map-support");
const util_1 = require("bdsx/util");
const krevent_1 = require("krevent");
const proc_1 = require("./proc");
const raknet_1 = require("./raknet");
class NetworkHandler extends nativeclass_1.NativeClass {
    send(ni, packet, u) {
        common_1.abstract();
    }
    sendInternal(ni, packet, data) {
        common_1.abstract();
    }
    getConnectionFromId(ni) {
        common_1.abstract();
    }
}
exports.NetworkHandler = NetworkHandler;
(function (NetworkHandler) {
    class Connection extends nativeclass_1.NativeClass {
    }
    NetworkHandler.Connection = Connection;
})(NetworkHandler = exports.NetworkHandler || (exports.NetworkHandler = {}));
let ServerNetworkHandler$Client = class ServerNetworkHandler$Client extends nativeclass_1.NativeClass {
};
ServerNetworkHandler$Client = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ServerNetworkHandler$Client);
let ServerNetworkHandler = class ServerNetworkHandler extends nativeclass_1.NativeClass {
    _disconnectClient(client, b, message, d) {
        common_1.abstract();
    }
    _setMotd(motd, shown) {
        common_1.abstract();
    }
    disconnectClient(client, message = "disconnectionScreen.disconnected") {
        const _message = new pointer_1.CxxStringWrapper(true);
        _message[nativetype_1.NativeType.ctor]();
        _message.value = message;
        this._disconnectClient(client, 0, _message, 0);
        _message[nativetype_1.NativeType.dtor]();
    }
    setMotd(motd) {
        const _motd = new pointer_1.CxxStringWrapper(true);
        _motd[nativetype_1.NativeType.ctor]();
        _motd.value = motd;
        this._setMotd(_motd, true);
        _motd[nativetype_1.NativeType.dtor]();
    }
};
ServerNetworkHandler = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ServerNetworkHandler);
exports.ServerNetworkHandler = ServerNetworkHandler;
const identifiers = new hashset_1.HashSet();
const closeEvTarget = new krevent_1.Event();
let NetworkIdentifier = NetworkIdentifier_1 = class NetworkIdentifier extends nativeclass_1.NativeClass {
    constructor(allocate) {
        super(allocate);
    }
    assignTo(target) {
        dll_1.dll.vcruntime140.memcpy(target, this, NetworkHandler[nativeclass_1.NativeClass.contentSize]);
    }
    equals(other) {
        common_1.abstract();
    }
    hash() {
        common_1.abstract();
    }
    getActor() {
        common_1.abstract();
    }
    getAddress() {
        const idx = this.address.GetSystemIndex();
        const rakpeer = exports.networkHandler.instance.peer;
        return rakpeer.GetSystemAddressFromIndex(idx).toString();
    }
    toString() {
        return this.getAddress();
    }
    static fromPointer(ptr) {
        return identifiers.get(ptr.as(NetworkIdentifier_1));
    }
    static [makefunc_1.makefunc.np2js](ptr) {
        let ni = identifiers.get(ptr);
        if (ni)
            return ni;
        ni = new NetworkIdentifier_1(true);
        ni.copyFrom(ptr, NetworkIdentifier_1[nativetype_1.NativeType.size]);
        identifiers.add(ni);
        return ni;
    }
    static all() {
        return identifiers.values();
    }
};
NetworkIdentifier.close = closeEvTarget;
tslib_1.__decorate([
    nativeclass_1.nativeField(raknet_1.RakNet.AddressOrGUID)
], NetworkIdentifier.prototype, "address", void 0);
NetworkIdentifier = NetworkIdentifier_1 = tslib_1.__decorate([
    nativeclass_1.nativeClass()
], NetworkIdentifier);
exports.NetworkIdentifier = NetworkIdentifier;
proc_1.procHacker.hookingRawWithCallOriginal('NetworkHandler::onConnectionClosed#1', makefunc_1.makefunc.np((handler, ni, msg) => {
    try {
        closeEvTarget.fire(ni);
        util_1._tickCallback();
    }
    catch (err) {
        source_map_support_1.remapAndPrintError(err);
    }
    // ni is used after onConnectionClosed. on some message processings.
    // timeout for avoiding the re-allocation
    setTimeout(() => {
        identifiers.delete(ni);
    }, 3000);
}, makefunc_1.RawTypeId.Void, null, NetworkHandler, NetworkIdentifier, pointer_1.CxxStringWrapper), [assembler_1.Register.rcx, assembler_1.Register.rdx, assembler_1.Register.r8, assembler_1.Register.r9], []);
