// bdsx/bds/packet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPacketRaw = exports.createPacket = exports.PacketSharedPtr = exports.Packet = exports.ExtendedStreamReadResult = exports.StreamReadResult = exports.PacketReadResult = void 0;
const tslib_1 = require("tslib");
const common_1 = require("bdsx/common");
const makefunc_1 = require("bdsx/makefunc");
const nativeclass_1 = require("bdsx/nativeclass");
const sharedpointer_1 = require("bdsx/sharedpointer");
const nativetype_1 = require("../nativetype");
const proc_1 = require("./proc");
// export interface PacketType<T> extends StructureType<T>
// {
//     readonly ID:number;
// }
exports.PacketReadResult = nativetype_1.uint32_t.extends({
    PacketReadNoError: 0,
    PacketReadError: 1,
});
exports.StreamReadResult = nativetype_1.int32_t.extends({
    Disconnect: 0,
    Pass: 1,
    Warning: 2,
    Ignore: 0x7f,
});
let ExtendedStreamReadResult = class ExtendedStreamReadResult extends nativeclass_1.NativeClass {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(exports.StreamReadResult)
], ExtendedStreamReadResult.prototype, "streamReadResult", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.int32_t)
], ExtendedStreamReadResult.prototype, "dummy", void 0);
ExtendedStreamReadResult = tslib_1.__decorate([
    nativeclass_1.nativeClass()
], ExtendedStreamReadResult);
exports.ExtendedStreamReadResult = ExtendedStreamReadResult;
const sharedptr_of_packet = Symbol('sharedptr');
let Packet = class Packet extends nativeclass_1.MantleClass {
    /**
     * @deprecated use packet.destruct();
     */
    destructor() {
        common_1.abstract();
    }
    getId() {
        common_1.abstract();
    }
    getName(name) {
        common_1.abstract();
    }
    write(stream) {
        common_1.abstract();
    }
    read(stream) {
        common_1.abstract();
    }
    readExtended(read, stream) {
        common_1.abstract();
    }
    /**
     * same with target.send
     */
    sendTo(target, unknownarg) {
        common_1.abstract();
    }
    dispose() {
        this[sharedptr_of_packet].dispose();
        this[sharedptr_of_packet] = null;
    }
    static create() {
        const id = this.ID;
        if (id === undefined)
            throw Error('Packet class is abstract, please use named class instead (ex. LoginPacket)');
        const cls = sharedpointer_1.SharedPtr.make(this);
        const sharedptr = new cls(true);
        exports.createPacketRaw(sharedptr, id);
        const packet = sharedptr.p;
        if (packet === null)
            throw Error(`${this.name} is not created`);
        packet[sharedptr_of_packet] = sharedptr;
        return packet;
    }
};
Packet = tslib_1.__decorate([
    nativeclass_1.nativeClass(0x30)
], Packet);
exports.Packet = Packet;
exports.PacketSharedPtr = sharedpointer_1.SharedPtr.make(Packet);
/**
 * @deprecated use *Packet.create() instead
 */
function createPacket(packetId) {
    const p = new exports.PacketSharedPtr(true);
    exports.createPacketRaw(p, packetId);
    return new sharedpointer_1.SharedPointer(p);
}
exports.createPacket = createPacket;
exports.createPacketRaw = proc_1.procHacker.js("MinecraftPackets::createPacket", exports.PacketSharedPtr, null, exports.PacketSharedPtr, makefunc_1.RawTypeId.Int32);
