// bdsx/bds/packets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerClosePacket = exports.ContainerOpenPacket = exports.RespawnPacket = exports.AnimatePacket = exports.SetSpawnPositionPacket = exports.SetHealthPacket = exports.SetActorLinkPacket = exports.SetActorMotionPacket = exports.SetActorDataPacket = exports.HurtArmorPacket = exports.EntityFallPacket = exports.PlayerActionPacket = exports.ActorPickRequestPacket = exports.BlockPickRequestPacket = exports.InteractPacket = exports.MobArmorEquipmentPacket = exports.MobEquipmentPacket = exports.InventoryTransactionPacket = exports.UpdateAttributesPacket = exports.AttributeData = exports.MobEffectPacket = exports.EntityEventPacket = exports.ActorEventPacket = exports.BlockEventPacket = exports.LevelEventPacket = exports.LevelSoundEventPacketV1 = exports.TickSyncPacket = exports.AddPaintingPacket = exports.UpdateBlockPacket = exports.RiderJumpPacket = exports.MovePlayerPacket = exports.MoveActorAbsolutePacket = exports.TakeItemActorPacket = exports.AddItemActorPacket = exports.RemoveActorPacket = exports.AddActorPacket = exports.AddPlayerPacket = exports.StartGamePacket = exports.LevelSettings = exports.SetTimePacket = exports.TextPacket = exports.ResourcePackClientResponsePacket = exports.ResourcePacksStackPacket = exports.ResourcePacksInfoPacket = exports.DisconnectPacket = exports.ClientToServerHandshakePacket = exports.ServerToClientHandshakePacket = exports.PlayStatusPacket = exports.LoginPacket = exports.NetworkBlockPosition = void 0;
exports.PhotoTransferPacket = exports.NpcRequestPacket = exports.BookEditPacket = exports.SetLastHurtByPacket = exports.WSConnectPacket = exports.SubClientLoginPacket = exports.PlayerSkinPacket = exports.PurchaseReceiptPacket = exports.ShowStoreOfferPacket = exports.StructureBlockUpdatePacket = exports.AddBehaviorTreePacket = exports.SetTitlePacket = exports.StopSoundPacket = exports.PlaySoundPacket = exports.TransferPacket = exports.ResourcePackChunkRequestPacket = exports.ResourcePackChunkDataPacket = exports.ResourcePackDataInfoPacket = exports.CommandOutputPacket = exports.CommandBlockUpdatePacket = exports.CommandRequestPacket = exports.AvailableCommandsPacket = exports.ShowCreditsPacket = exports.BossEventPacket = exports.CameraPacket = exports.GameRulesChangedPacket = exports.ItemFrameDropItemPacket = exports.ChunkRadiusUpdatedPacket = exports.RequestChunkRadiusPacket = exports.MapInfoRequestPacket = exports.MapItemDataPacket = exports.SpawnExperienceOrbPacket = exports.TelemetryEventPacket = exports.SimpleEventPacket = exports.PlayerListPacket = exports.SetPlayerGameTypePacket = exports.ChangeDimensionPacket = exports.SetDifficultyPacket = exports.SetCommandsEnabledPacket = exports.LevelChunkPacket = exports.PlayerInputPacket = exports.BlockActorDataPacket = exports.AdventureSettingsPacket = exports.GuiDataPickItemPacket = exports.CraftingEventPacket = exports.CraftingDataPacket = exports.ContainerSetDataPacket = exports.InventorySlotPacket = exports.InventoryContentPacket = exports.PlayerHotbarPacket = void 0;
exports.EmoteListPacket = exports.UpdatePlayerGameTypePacket = exports.CodeBuilderPacket = exports.PlayerArmorDamagePacket = exports.ItemStackResponse = exports.ItemStackRequest = exports.PlayerEnchantOptionsPacket = exports.CreativeContentPacket = exports.PlayerAuthInputPacket = exports.NetworkSettingsPacket = exports.CompletedUsingItemPacket = exports.AnvilDamagePacket = exports.SettingsCommandPacket = exports.MultiplayerSettingsPacket = exports.EmotePacket = exports.EducationSettingsPacket = exports.ClientCacheMissResponsePacket = exports.ClientCacheBlobStatusPacket = exports.StructureTemplateDataExportPacket = exports.StructureTemplateDataRequestPacket = exports.MapCreateLockedCopy = exports.OnScreenTextureAnimationPacket = exports.ClientCacheStatusPacket = exports.RemoveEntityPacket = exports.LecternUpdatePacket = exports.LevelEventGenericPacket = exports.LevelSoundEventPacket = exports.BiomeDefinitionList = exports.NetworkChunkPublisherUpdatePacket = exports.LevelSoundEventPacketV2 = exports.AvailableActorIdentifiersPacket = exports.SpawnParticleEffect = exports.ScriptCustomEventPacket = exports.NetworkStackLatencyPacket = exports.UpdateSoftEnumPacket = exports.SetLocalPlayerAsInitializedPacket = exports.SetScoreboardIdentityPacket = exports.MoveActorDeltaPacket = exports.UpdateBlockPacketSynced = exports.LabTablePacket = exports.SetScorePacket = exports.SetDisplayObjectivePacket = exports.RemoveObjectivePacket = exports.SetDefaultGameTypePacket = exports.ShowProfilePacket = exports.ServerSettingsResponsePacket = exports.ServerSettingsRequestPacket = exports.ModalFormResponsePacket = exports.ModalFormRequestPacket = exports.ShowModalFormPacket = void 0;
exports.PacketIdToType = exports.FilterTextPacket = exports.ItemComponentPacket = exports.CorrectPlayerMovePredictionPacket = exports.PlayerFogPacket = exports.CameraShakePacket = exports.AnimateEntityPacket = exports.MotionPredictionHintsPacket = exports.PacketViolationWarningPacket = exports.DebugInfoPacket = exports.PositionTrackingDBClientRequest = exports.PositionTrackingDBServerBroadcast = void 0;
const tslib_1 = require("tslib");
const cxxvector_1 = require("bdsx/cxxvector");
const nativeclass_1 = require("bdsx/nativeclass");
const nativetype_1 = require("bdsx/nativetype");
const actor_1 = require("./actor");
const blockpos_1 = require("./blockpos");
const connreq_1 = require("./connreq");
const hashedstring_1 = require("./hashedstring");
const packet_1 = require("./packet");
/** @deprecated use BlockPos instead */
exports.NetworkBlockPosition = blockpos_1.BlockPos;
let LoginPacket = class LoginPacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.int32_t, 0x30)
], LoginPacket.prototype, "protocol", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(connreq_1.ConnectionRequest.ref(), 0x38)
], LoginPacket.prototype, "connreq", void 0);
LoginPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], LoginPacket);
exports.LoginPacket = LoginPacket;
let PlayStatusPacket = class PlayStatusPacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.int32_t)
], PlayStatusPacket.prototype, "status", void 0);
PlayStatusPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], PlayStatusPacket);
exports.PlayStatusPacket = PlayStatusPacket;
let ServerToClientHandshakePacket = class ServerToClientHandshakePacket extends packet_1.Packet {
};
ServerToClientHandshakePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ServerToClientHandshakePacket);
exports.ServerToClientHandshakePacket = ServerToClientHandshakePacket;
let ClientToServerHandshakePacket = class ClientToServerHandshakePacket extends packet_1.Packet {
};
ClientToServerHandshakePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ClientToServerHandshakePacket);
exports.ClientToServerHandshakePacket = ClientToServerHandshakePacket;
let DisconnectPacket = class DisconnectPacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.CxxString, 0x38)
], DisconnectPacket.prototype, "message", void 0);
DisconnectPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], DisconnectPacket);
exports.DisconnectPacket = DisconnectPacket;
let ResourcePacksInfoPacket = class ResourcePacksInfoPacket extends packet_1.Packet {
};
ResourcePacksInfoPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ResourcePacksInfoPacket);
exports.ResourcePacksInfoPacket = ResourcePacksInfoPacket;
let ResourcePacksStackPacket = class ResourcePacksStackPacket extends packet_1.Packet {
};
ResourcePacksStackPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ResourcePacksStackPacket);
exports.ResourcePacksStackPacket = ResourcePacksStackPacket;
let ResourcePackClientResponsePacket = class ResourcePackClientResponsePacket extends packet_1.Packet {
};
ResourcePackClientResponsePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ResourcePackClientResponsePacket);
exports.ResourcePackClientResponsePacket = ResourcePackClientResponsePacket;
let TextPacket = class TextPacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint8_t)
], TextPacket.prototype, "type", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.CxxString)
], TextPacket.prototype, "name", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.CxxString)
], TextPacket.prototype, "message", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.bool_t, 0x90)
], TextPacket.prototype, "needsTranslation", void 0);
TextPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], TextPacket);
exports.TextPacket = TextPacket;
let SetTimePacket = class SetTimePacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.int32_t)
], SetTimePacket.prototype, "time", void 0);
SetTimePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], SetTimePacket);
exports.SetTimePacket = SetTimePacket;
let LevelSettings = class LevelSettings extends nativeclass_1.MantleClass {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.int32_t)
], LevelSettings.prototype, "seed", void 0);
LevelSettings = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], LevelSettings);
exports.LevelSettings = LevelSettings;
let StartGamePacket = class StartGamePacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(LevelSettings)
], StartGamePacket.prototype, "settings", void 0);
StartGamePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], StartGamePacket);
exports.StartGamePacket = StartGamePacket;
let AddPlayerPacket = class AddPlayerPacket extends packet_1.Packet {
};
AddPlayerPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], AddPlayerPacket);
exports.AddPlayerPacket = AddPlayerPacket;
let AddActorPacket = class AddActorPacket extends packet_1.Packet {
};
AddActorPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], AddActorPacket);
exports.AddActorPacket = AddActorPacket;
let RemoveActorPacket = class RemoveActorPacket extends packet_1.Packet {
};
RemoveActorPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], RemoveActorPacket);
exports.RemoveActorPacket = RemoveActorPacket;
let AddItemActorPacket = class AddItemActorPacket extends packet_1.Packet {
};
AddItemActorPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], AddItemActorPacket);
exports.AddItemActorPacket = AddItemActorPacket;
let TakeItemActorPacket = class TakeItemActorPacket extends packet_1.Packet {
};
TakeItemActorPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], TakeItemActorPacket);
exports.TakeItemActorPacket = TakeItemActorPacket;
let MoveActorAbsolutePacket = class MoveActorAbsolutePacket extends packet_1.Packet {
};
MoveActorAbsolutePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], MoveActorAbsolutePacket);
exports.MoveActorAbsolutePacket = MoveActorAbsolutePacket;
let MovePlayerPacket = class MovePlayerPacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(actor_1.ActorRuntimeID)
], MovePlayerPacket.prototype, "actorId", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(blockpos_1.Vec3)
], MovePlayerPacket.prototype, "pos", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.float32_t)
], MovePlayerPacket.prototype, "pitch", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.float32_t)
], MovePlayerPacket.prototype, "yaw", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.float32_t)
], MovePlayerPacket.prototype, "headYaw", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint8_t)
], MovePlayerPacket.prototype, "mode", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.bool_t)
], MovePlayerPacket.prototype, "onGround", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(actor_1.ActorRuntimeID)
], MovePlayerPacket.prototype, "ridingActorId", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.int32_t)
], MovePlayerPacket.prototype, "teleportCause", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.int32_t)
], MovePlayerPacket.prototype, "teleportItem", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.bin64_t)
], MovePlayerPacket.prototype, "tick", void 0);
MovePlayerPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], MovePlayerPacket);
exports.MovePlayerPacket = MovePlayerPacket;
let RiderJumpPacket = class RiderJumpPacket extends packet_1.Packet {
};
RiderJumpPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], RiderJumpPacket);
exports.RiderJumpPacket = RiderJumpPacket;
let UpdateBlockPacket = class UpdateBlockPacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(blockpos_1.BlockPos)
], UpdateBlockPacket.prototype, "blockPos", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint32_t)
], UpdateBlockPacket.prototype, "blockRuntimeId", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint8_t)
], UpdateBlockPacket.prototype, "flags", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint32_t)
], UpdateBlockPacket.prototype, "dataLayerId", void 0);
UpdateBlockPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], UpdateBlockPacket);
exports.UpdateBlockPacket = UpdateBlockPacket;
let AddPaintingPacket = class AddPaintingPacket extends packet_1.Packet {
};
AddPaintingPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], AddPaintingPacket);
exports.AddPaintingPacket = AddPaintingPacket;
let TickSyncPacket = class TickSyncPacket extends packet_1.Packet {
};
TickSyncPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], TickSyncPacket);
exports.TickSyncPacket = TickSyncPacket;
let LevelSoundEventPacketV1 = class LevelSoundEventPacketV1 extends packet_1.Packet {
};
LevelSoundEventPacketV1 = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], LevelSoundEventPacketV1);
exports.LevelSoundEventPacketV1 = LevelSoundEventPacketV1;
let LevelEventPacket = class LevelEventPacket extends packet_1.Packet {
};
LevelEventPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], LevelEventPacket);
exports.LevelEventPacket = LevelEventPacket;
let BlockEventPacket = class BlockEventPacket extends packet_1.Packet {
};
BlockEventPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], BlockEventPacket);
exports.BlockEventPacket = BlockEventPacket;
let ActorEventPacket = class ActorEventPacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(actor_1.ActorRuntimeID)
], ActorEventPacket.prototype, "actorId", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint8_t)
], ActorEventPacket.prototype, "event", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.int32_t)
], ActorEventPacket.prototype, "data", void 0);
ActorEventPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ActorEventPacket);
exports.ActorEventPacket = ActorEventPacket;
/** @deprecated use ActorEventPacket, matching to official name */
exports.EntityEventPacket = ActorEventPacket;
let MobEffectPacket = class MobEffectPacket extends packet_1.Packet {
};
MobEffectPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], MobEffectPacket);
exports.MobEffectPacket = MobEffectPacket;
let AttributeData = class AttributeData extends nativeclass_1.NativeClass {
    [nativetype_1.NativeType.ctor]() {
        this.min = 0;
        this.max = 0;
        this.current = 0;
        this.default = 0;
    }
};
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.float32_t)
], AttributeData.prototype, "current", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.float32_t)
], AttributeData.prototype, "min", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.float32_t)
], AttributeData.prototype, "max", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.float32_t)
], AttributeData.prototype, "default", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(hashedstring_1.HashedString)
], AttributeData.prototype, "name", void 0);
AttributeData = tslib_1.__decorate([
    nativeclass_1.nativeClass(0x40)
], AttributeData);
exports.AttributeData = AttributeData;
let UpdateAttributesPacket = class UpdateAttributesPacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(actor_1.ActorRuntimeID)
], UpdateAttributesPacket.prototype, "actorId", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(cxxvector_1.CxxVector.make(AttributeData))
], UpdateAttributesPacket.prototype, "attributes", void 0);
UpdateAttributesPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], UpdateAttributesPacket);
exports.UpdateAttributesPacket = UpdateAttributesPacket;
let InventoryTransactionPacket = class InventoryTransactionPacket extends packet_1.Packet {
};
InventoryTransactionPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], InventoryTransactionPacket);
exports.InventoryTransactionPacket = InventoryTransactionPacket;
let MobEquipmentPacket = class MobEquipmentPacket extends packet_1.Packet {
};
MobEquipmentPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], MobEquipmentPacket);
exports.MobEquipmentPacket = MobEquipmentPacket;
let MobArmorEquipmentPacket = class MobArmorEquipmentPacket extends packet_1.Packet {
};
MobArmorEquipmentPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], MobArmorEquipmentPacket);
exports.MobArmorEquipmentPacket = MobArmorEquipmentPacket;
let InteractPacket = class InteractPacket extends packet_1.Packet {
};
InteractPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], InteractPacket);
exports.InteractPacket = InteractPacket;
let BlockPickRequestPacket = class BlockPickRequestPacket extends packet_1.Packet {
};
BlockPickRequestPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], BlockPickRequestPacket);
exports.BlockPickRequestPacket = BlockPickRequestPacket;
let ActorPickRequestPacket = class ActorPickRequestPacket extends packet_1.Packet {
};
ActorPickRequestPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ActorPickRequestPacket);
exports.ActorPickRequestPacket = ActorPickRequestPacket;
let PlayerActionPacket = class PlayerActionPacket extends packet_1.Packet {
};
PlayerActionPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], PlayerActionPacket);
exports.PlayerActionPacket = PlayerActionPacket;
let EntityFallPacket = class EntityFallPacket extends packet_1.Packet {
};
EntityFallPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], EntityFallPacket);
exports.EntityFallPacket = EntityFallPacket;
let HurtArmorPacket = class HurtArmorPacket extends packet_1.Packet {
};
HurtArmorPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], HurtArmorPacket);
exports.HurtArmorPacket = HurtArmorPacket;
let SetActorDataPacket = class SetActorDataPacket extends packet_1.Packet {
};
SetActorDataPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], SetActorDataPacket);
exports.SetActorDataPacket = SetActorDataPacket;
let SetActorMotionPacket = class SetActorMotionPacket extends packet_1.Packet {
};
SetActorMotionPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], SetActorMotionPacket);
exports.SetActorMotionPacket = SetActorMotionPacket;
let SetActorLinkPacket = class SetActorLinkPacket extends packet_1.Packet {
};
SetActorLinkPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], SetActorLinkPacket);
exports.SetActorLinkPacket = SetActorLinkPacket;
let SetHealthPacket = class SetHealthPacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint8_t)
], SetHealthPacket.prototype, "health", void 0);
SetHealthPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], SetHealthPacket);
exports.SetHealthPacket = SetHealthPacket;
let SetSpawnPositionPacket = class SetSpawnPositionPacket extends packet_1.Packet {
};
SetSpawnPositionPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], SetSpawnPositionPacket);
exports.SetSpawnPositionPacket = SetSpawnPositionPacket;
let AnimatePacket = class AnimatePacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(actor_1.ActorRuntimeID)
], AnimatePacket.prototype, "actorId", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint8_t)
], AnimatePacket.prototype, "action", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.float32_t)
], AnimatePacket.prototype, "unknown", void 0);
AnimatePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], AnimatePacket);
exports.AnimatePacket = AnimatePacket;
let RespawnPacket = class RespawnPacket extends packet_1.Packet {
};
RespawnPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], RespawnPacket);
exports.RespawnPacket = RespawnPacket;
let ContainerOpenPacket = class ContainerOpenPacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint8_t)
], ContainerOpenPacket.prototype, "windowId", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.int8_t)
], ContainerOpenPacket.prototype, "type", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(blockpos_1.BlockPos)
], ContainerOpenPacket.prototype, "pos", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.bin64_t)
], ContainerOpenPacket.prototype, "entityUniqueId", void 0);
ContainerOpenPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ContainerOpenPacket);
exports.ContainerOpenPacket = ContainerOpenPacket;
let ContainerClosePacket = class ContainerClosePacket extends packet_1.Packet {
};
ContainerClosePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ContainerClosePacket);
exports.ContainerClosePacket = ContainerClosePacket;
let PlayerHotbarPacket = class PlayerHotbarPacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint32_t)
], PlayerHotbarPacket.prototype, "selectedSlot", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.bool_t)
], PlayerHotbarPacket.prototype, "selectHotbarSlot", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint8_t)
], PlayerHotbarPacket.prototype, "windowId", void 0);
PlayerHotbarPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], PlayerHotbarPacket);
exports.PlayerHotbarPacket = PlayerHotbarPacket;
let InventoryContentPacket = class InventoryContentPacket extends packet_1.Packet {
};
InventoryContentPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], InventoryContentPacket);
exports.InventoryContentPacket = InventoryContentPacket;
let InventorySlotPacket = class InventorySlotPacket extends packet_1.Packet {
};
InventorySlotPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], InventorySlotPacket);
exports.InventorySlotPacket = InventorySlotPacket;
let ContainerSetDataPacket = class ContainerSetDataPacket extends packet_1.Packet {
};
ContainerSetDataPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ContainerSetDataPacket);
exports.ContainerSetDataPacket = ContainerSetDataPacket;
let CraftingDataPacket = class CraftingDataPacket extends packet_1.Packet {
};
CraftingDataPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], CraftingDataPacket);
exports.CraftingDataPacket = CraftingDataPacket;
let CraftingEventPacket = class CraftingEventPacket extends packet_1.Packet {
};
CraftingEventPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], CraftingEventPacket);
exports.CraftingEventPacket = CraftingEventPacket;
let GuiDataPickItemPacket = class GuiDataPickItemPacket extends packet_1.Packet {
};
GuiDataPickItemPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], GuiDataPickItemPacket);
exports.GuiDataPickItemPacket = GuiDataPickItemPacket;
let AdventureSettingsPacket = class AdventureSettingsPacket extends packet_1.Packet {
};
AdventureSettingsPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], AdventureSettingsPacket);
exports.AdventureSettingsPacket = AdventureSettingsPacket;
let BlockActorDataPacket = class BlockActorDataPacket extends packet_1.Packet {
};
BlockActorDataPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], BlockActorDataPacket);
exports.BlockActorDataPacket = BlockActorDataPacket;
let PlayerInputPacket = class PlayerInputPacket extends packet_1.Packet {
};
PlayerInputPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], PlayerInputPacket);
exports.PlayerInputPacket = PlayerInputPacket;
let LevelChunkPacket = class LevelChunkPacket extends packet_1.Packet {
};
LevelChunkPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], LevelChunkPacket);
exports.LevelChunkPacket = LevelChunkPacket;
let SetCommandsEnabledPacket = class SetCommandsEnabledPacket extends packet_1.Packet {
};
SetCommandsEnabledPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], SetCommandsEnabledPacket);
exports.SetCommandsEnabledPacket = SetCommandsEnabledPacket;
let SetDifficultyPacket = class SetDifficultyPacket extends packet_1.Packet {
};
SetDifficultyPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], SetDifficultyPacket);
exports.SetDifficultyPacket = SetDifficultyPacket;
let ChangeDimensionPacket = class ChangeDimensionPacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint32_t)
], ChangeDimensionPacket.prototype, "dimensionId", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.float32_t)
], ChangeDimensionPacket.prototype, "x", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.float32_t)
], ChangeDimensionPacket.prototype, "y", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.float32_t)
], ChangeDimensionPacket.prototype, "z", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.bool_t)
], ChangeDimensionPacket.prototype, "respawn", void 0);
ChangeDimensionPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ChangeDimensionPacket);
exports.ChangeDimensionPacket = ChangeDimensionPacket;
let SetPlayerGameTypePacket = class SetPlayerGameTypePacket extends packet_1.Packet {
};
SetPlayerGameTypePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], SetPlayerGameTypePacket);
exports.SetPlayerGameTypePacket = SetPlayerGameTypePacket;
let PlayerListPacket = class PlayerListPacket extends packet_1.Packet {
};
PlayerListPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], PlayerListPacket);
exports.PlayerListPacket = PlayerListPacket;
let SimpleEventPacket = class SimpleEventPacket extends packet_1.Packet {
};
SimpleEventPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], SimpleEventPacket);
exports.SimpleEventPacket = SimpleEventPacket;
let TelemetryEventPacket = class TelemetryEventPacket extends packet_1.Packet {
};
TelemetryEventPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], TelemetryEventPacket);
exports.TelemetryEventPacket = TelemetryEventPacket;
let SpawnExperienceOrbPacket = class SpawnExperienceOrbPacket extends packet_1.Packet {
};
SpawnExperienceOrbPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], SpawnExperienceOrbPacket);
exports.SpawnExperienceOrbPacket = SpawnExperienceOrbPacket;
let MapItemDataPacket = class MapItemDataPacket extends packet_1.Packet {
};
MapItemDataPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], MapItemDataPacket);
exports.MapItemDataPacket = MapItemDataPacket;
let MapInfoRequestPacket = class MapInfoRequestPacket extends packet_1.Packet {
};
MapInfoRequestPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], MapInfoRequestPacket);
exports.MapInfoRequestPacket = MapInfoRequestPacket;
let RequestChunkRadiusPacket = class RequestChunkRadiusPacket extends packet_1.Packet {
};
RequestChunkRadiusPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], RequestChunkRadiusPacket);
exports.RequestChunkRadiusPacket = RequestChunkRadiusPacket;
let ChunkRadiusUpdatedPacket = class ChunkRadiusUpdatedPacket extends packet_1.Packet {
};
ChunkRadiusUpdatedPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ChunkRadiusUpdatedPacket);
exports.ChunkRadiusUpdatedPacket = ChunkRadiusUpdatedPacket;
let ItemFrameDropItemPacket = class ItemFrameDropItemPacket extends packet_1.Packet {
};
ItemFrameDropItemPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ItemFrameDropItemPacket);
exports.ItemFrameDropItemPacket = ItemFrameDropItemPacket;
let GameRulesChangedPacket = class GameRulesChangedPacket extends packet_1.Packet {
};
GameRulesChangedPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], GameRulesChangedPacket);
exports.GameRulesChangedPacket = GameRulesChangedPacket;
let CameraPacket = class CameraPacket extends packet_1.Packet {
};
CameraPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], CameraPacket);
exports.CameraPacket = CameraPacket;
let BossEventPacket = class BossEventPacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.bin64_t)
], BossEventPacket.prototype, "unknown", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.bin64_t)
], BossEventPacket.prototype, "entityUniqueId", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.bin64_t)
], BossEventPacket.prototype, "unknown2", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint32_t)
], BossEventPacket.prototype, "type", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.CxxString)
], BossEventPacket.prototype, "title", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.float32_t)
], BossEventPacket.prototype, "healthPercent", void 0);
BossEventPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], BossEventPacket);
exports.BossEventPacket = BossEventPacket;
let ShowCreditsPacket = class ShowCreditsPacket extends packet_1.Packet {
};
ShowCreditsPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ShowCreditsPacket);
exports.ShowCreditsPacket = ShowCreditsPacket;
let AvailableCommandsPacket = class AvailableCommandsPacket extends packet_1.Packet {
};
AvailableCommandsPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], AvailableCommandsPacket);
exports.AvailableCommandsPacket = AvailableCommandsPacket;
let CommandRequestPacket = class CommandRequestPacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.CxxString)
], CommandRequestPacket.prototype, "command", void 0);
CommandRequestPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], CommandRequestPacket);
exports.CommandRequestPacket = CommandRequestPacket;
let CommandBlockUpdatePacket = class CommandBlockUpdatePacket extends packet_1.Packet {
};
CommandBlockUpdatePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], CommandBlockUpdatePacket);
exports.CommandBlockUpdatePacket = CommandBlockUpdatePacket;
let CommandOutputPacket = class CommandOutputPacket extends packet_1.Packet {
};
CommandOutputPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], CommandOutputPacket);
exports.CommandOutputPacket = CommandOutputPacket;
let ResourcePackDataInfoPacket = class ResourcePackDataInfoPacket extends packet_1.Packet {
};
ResourcePackDataInfoPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ResourcePackDataInfoPacket);
exports.ResourcePackDataInfoPacket = ResourcePackDataInfoPacket;
let ResourcePackChunkDataPacket = class ResourcePackChunkDataPacket extends packet_1.Packet {
};
ResourcePackChunkDataPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ResourcePackChunkDataPacket);
exports.ResourcePackChunkDataPacket = ResourcePackChunkDataPacket;
let ResourcePackChunkRequestPacket = class ResourcePackChunkRequestPacket extends packet_1.Packet {
};
ResourcePackChunkRequestPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ResourcePackChunkRequestPacket);
exports.ResourcePackChunkRequestPacket = ResourcePackChunkRequestPacket;
let TransferPacket = class TransferPacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.CxxString)
], TransferPacket.prototype, "address", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint16_t)
], TransferPacket.prototype, "port", void 0);
TransferPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], TransferPacket);
exports.TransferPacket = TransferPacket;
let PlaySoundPacket = class PlaySoundPacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.CxxString)
], PlaySoundPacket.prototype, "soundName", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(blockpos_1.BlockPos)
], PlaySoundPacket.prototype, "pos", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.float32_t)
], PlaySoundPacket.prototype, "volume", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.float32_t)
], PlaySoundPacket.prototype, "pitch", void 0);
PlaySoundPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], PlaySoundPacket);
exports.PlaySoundPacket = PlaySoundPacket;
let StopSoundPacket = class StopSoundPacket extends packet_1.Packet {
};
StopSoundPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], StopSoundPacket);
exports.StopSoundPacket = StopSoundPacket;
let SetTitlePacket = class SetTitlePacket extends packet_1.Packet {
};
SetTitlePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], SetTitlePacket);
exports.SetTitlePacket = SetTitlePacket;
let AddBehaviorTreePacket = class AddBehaviorTreePacket extends packet_1.Packet {
};
AddBehaviorTreePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], AddBehaviorTreePacket);
exports.AddBehaviorTreePacket = AddBehaviorTreePacket;
let StructureBlockUpdatePacket = class StructureBlockUpdatePacket extends packet_1.Packet {
};
StructureBlockUpdatePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], StructureBlockUpdatePacket);
exports.StructureBlockUpdatePacket = StructureBlockUpdatePacket;
let ShowStoreOfferPacket = class ShowStoreOfferPacket extends packet_1.Packet {
};
ShowStoreOfferPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ShowStoreOfferPacket);
exports.ShowStoreOfferPacket = ShowStoreOfferPacket;
let PurchaseReceiptPacket = class PurchaseReceiptPacket extends packet_1.Packet {
};
PurchaseReceiptPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], PurchaseReceiptPacket);
exports.PurchaseReceiptPacket = PurchaseReceiptPacket;
let PlayerSkinPacket = class PlayerSkinPacket extends packet_1.Packet {
};
PlayerSkinPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], PlayerSkinPacket);
exports.PlayerSkinPacket = PlayerSkinPacket;
let SubClientLoginPacket = class SubClientLoginPacket extends packet_1.Packet {
};
SubClientLoginPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], SubClientLoginPacket);
exports.SubClientLoginPacket = SubClientLoginPacket;
let WSConnectPacket = class WSConnectPacket extends packet_1.Packet {
};
WSConnectPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], WSConnectPacket);
exports.WSConnectPacket = WSConnectPacket;
let SetLastHurtByPacket = class SetLastHurtByPacket extends packet_1.Packet {
};
SetLastHurtByPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], SetLastHurtByPacket);
exports.SetLastHurtByPacket = SetLastHurtByPacket;
let BookEditPacket = class BookEditPacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint8_t, 0x30)
], BookEditPacket.prototype, "type", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint8_t, 0x34)
], BookEditPacket.prototype, "inventorySlot", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint8_t, 0x38)
], BookEditPacket.prototype, "pageNumber", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint8_t, 0x3c)
], BookEditPacket.prototype, "secondaryPageNumber", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.CxxString, 0x40)
], BookEditPacket.prototype, "text", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.CxxString, 0x60)
], BookEditPacket.prototype, "author", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.CxxString, 0x80)
], BookEditPacket.prototype, "xuid", void 0);
BookEditPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], BookEditPacket);
exports.BookEditPacket = BookEditPacket;
let NpcRequestPacket = class NpcRequestPacket extends packet_1.Packet {
};
NpcRequestPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], NpcRequestPacket);
exports.NpcRequestPacket = NpcRequestPacket;
let PhotoTransferPacket = class PhotoTransferPacket extends packet_1.Packet {
};
PhotoTransferPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], PhotoTransferPacket);
exports.PhotoTransferPacket = PhotoTransferPacket;
let ShowModalFormPacket = class ShowModalFormPacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint32_t)
], ShowModalFormPacket.prototype, "id", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.CxxString)
], ShowModalFormPacket.prototype, "content", void 0);
ShowModalFormPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ShowModalFormPacket);
exports.ShowModalFormPacket = ShowModalFormPacket;
/** @deprecated use ShowModalFormPacket, matching to official name */
exports.ModalFormRequestPacket = ShowModalFormPacket;
let ModalFormResponsePacket = class ModalFormResponsePacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint32_t)
], ModalFormResponsePacket.prototype, "id", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.CxxString)
], ModalFormResponsePacket.prototype, "response", void 0);
ModalFormResponsePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ModalFormResponsePacket);
exports.ModalFormResponsePacket = ModalFormResponsePacket;
let ServerSettingsRequestPacket = class ServerSettingsRequestPacket extends packet_1.Packet {
};
ServerSettingsRequestPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ServerSettingsRequestPacket);
exports.ServerSettingsRequestPacket = ServerSettingsRequestPacket;
let ServerSettingsResponsePacket = class ServerSettingsResponsePacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint32_t)
], ServerSettingsResponsePacket.prototype, "id", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.CxxString)
], ServerSettingsResponsePacket.prototype, "content", void 0);
ServerSettingsResponsePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ServerSettingsResponsePacket);
exports.ServerSettingsResponsePacket = ServerSettingsResponsePacket;
let ShowProfilePacket = class ShowProfilePacket extends packet_1.Packet {
};
ShowProfilePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ShowProfilePacket);
exports.ShowProfilePacket = ShowProfilePacket;
let SetDefaultGameTypePacket = class SetDefaultGameTypePacket extends packet_1.Packet {
};
SetDefaultGameTypePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], SetDefaultGameTypePacket);
exports.SetDefaultGameTypePacket = SetDefaultGameTypePacket;
let RemoveObjectivePacket = class RemoveObjectivePacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.CxxString)
], RemoveObjectivePacket.prototype, "objectiveName", void 0);
RemoveObjectivePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], RemoveObjectivePacket);
exports.RemoveObjectivePacket = RemoveObjectivePacket;
let SetDisplayObjectivePacket = class SetDisplayObjectivePacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.CxxString)
], SetDisplayObjectivePacket.prototype, "displaySlot", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.CxxString)
], SetDisplayObjectivePacket.prototype, "objectiveName", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.CxxString)
], SetDisplayObjectivePacket.prototype, "displayName", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.CxxString)
], SetDisplayObjectivePacket.prototype, "criteriaName", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.int32_t)
], SetDisplayObjectivePacket.prototype, "sortOrder", void 0);
SetDisplayObjectivePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], SetDisplayObjectivePacket);
exports.SetDisplayObjectivePacket = SetDisplayObjectivePacket;
let SetScorePacket = class SetScorePacket extends packet_1.Packet {
};
SetScorePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], SetScorePacket);
exports.SetScorePacket = SetScorePacket;
let LabTablePacket = class LabTablePacket extends packet_1.Packet {
};
LabTablePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], LabTablePacket);
exports.LabTablePacket = LabTablePacket;
let UpdateBlockPacketSynced = class UpdateBlockPacketSynced extends packet_1.Packet {
};
UpdateBlockPacketSynced = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], UpdateBlockPacketSynced);
exports.UpdateBlockPacketSynced = UpdateBlockPacketSynced;
let MoveActorDeltaPacket = class MoveActorDeltaPacket extends packet_1.Packet {
};
MoveActorDeltaPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], MoveActorDeltaPacket);
exports.MoveActorDeltaPacket = MoveActorDeltaPacket;
let SetScoreboardIdentityPacket = class SetScoreboardIdentityPacket extends packet_1.Packet {
};
SetScoreboardIdentityPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], SetScoreboardIdentityPacket);
exports.SetScoreboardIdentityPacket = SetScoreboardIdentityPacket;
let SetLocalPlayerAsInitializedPacket = class SetLocalPlayerAsInitializedPacket extends packet_1.Packet {
};
SetLocalPlayerAsInitializedPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], SetLocalPlayerAsInitializedPacket);
exports.SetLocalPlayerAsInitializedPacket = SetLocalPlayerAsInitializedPacket;
let UpdateSoftEnumPacket = class UpdateSoftEnumPacket extends packet_1.Packet {
};
UpdateSoftEnumPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], UpdateSoftEnumPacket);
exports.UpdateSoftEnumPacket = UpdateSoftEnumPacket;
let NetworkStackLatencyPacket = class NetworkStackLatencyPacket extends packet_1.Packet {
};
NetworkStackLatencyPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], NetworkStackLatencyPacket);
exports.NetworkStackLatencyPacket = NetworkStackLatencyPacket;
let ScriptCustomEventPacket = class ScriptCustomEventPacket extends packet_1.Packet {
};
ScriptCustomEventPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ScriptCustomEventPacket);
exports.ScriptCustomEventPacket = ScriptCustomEventPacket;
let SpawnParticleEffect = class SpawnParticleEffect extends packet_1.Packet {
};
SpawnParticleEffect = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], SpawnParticleEffect);
exports.SpawnParticleEffect = SpawnParticleEffect;
let AvailableActorIdentifiersPacket = class AvailableActorIdentifiersPacket extends packet_1.Packet {
};
AvailableActorIdentifiersPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], AvailableActorIdentifiersPacket);
exports.AvailableActorIdentifiersPacket = AvailableActorIdentifiersPacket;
let LevelSoundEventPacketV2 = class LevelSoundEventPacketV2 extends packet_1.Packet {
};
LevelSoundEventPacketV2 = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], LevelSoundEventPacketV2);
exports.LevelSoundEventPacketV2 = LevelSoundEventPacketV2;
let NetworkChunkPublisherUpdatePacket = class NetworkChunkPublisherUpdatePacket extends packet_1.Packet {
};
NetworkChunkPublisherUpdatePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], NetworkChunkPublisherUpdatePacket);
exports.NetworkChunkPublisherUpdatePacket = NetworkChunkPublisherUpdatePacket;
let BiomeDefinitionList = class BiomeDefinitionList extends packet_1.Packet {
};
BiomeDefinitionList = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], BiomeDefinitionList);
exports.BiomeDefinitionList = BiomeDefinitionList;
let LevelSoundEventPacket = class LevelSoundEventPacket extends packet_1.Packet {
};
LevelSoundEventPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], LevelSoundEventPacket);
exports.LevelSoundEventPacket = LevelSoundEventPacket;
let LevelEventGenericPacket = class LevelEventGenericPacket extends packet_1.Packet {
};
LevelEventGenericPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], LevelEventGenericPacket);
exports.LevelEventGenericPacket = LevelEventGenericPacket;
let LecternUpdatePacket = class LecternUpdatePacket extends packet_1.Packet {
};
LecternUpdatePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], LecternUpdatePacket);
exports.LecternUpdatePacket = LecternUpdatePacket;
let RemoveEntityPacket = class RemoveEntityPacket extends packet_1.Packet {
};
RemoveEntityPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], RemoveEntityPacket);
exports.RemoveEntityPacket = RemoveEntityPacket;
let ClientCacheStatusPacket = class ClientCacheStatusPacket extends packet_1.Packet {
};
ClientCacheStatusPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ClientCacheStatusPacket);
exports.ClientCacheStatusPacket = ClientCacheStatusPacket;
let OnScreenTextureAnimationPacket = class OnScreenTextureAnimationPacket extends packet_1.Packet {
};
OnScreenTextureAnimationPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], OnScreenTextureAnimationPacket);
exports.OnScreenTextureAnimationPacket = OnScreenTextureAnimationPacket;
let MapCreateLockedCopy = class MapCreateLockedCopy extends packet_1.Packet {
};
MapCreateLockedCopy = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], MapCreateLockedCopy);
exports.MapCreateLockedCopy = MapCreateLockedCopy;
let StructureTemplateDataRequestPacket = class StructureTemplateDataRequestPacket extends packet_1.Packet {
};
StructureTemplateDataRequestPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], StructureTemplateDataRequestPacket);
exports.StructureTemplateDataRequestPacket = StructureTemplateDataRequestPacket;
let StructureTemplateDataExportPacket = class StructureTemplateDataExportPacket extends packet_1.Packet {
};
StructureTemplateDataExportPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], StructureTemplateDataExportPacket);
exports.StructureTemplateDataExportPacket = StructureTemplateDataExportPacket;
let ClientCacheBlobStatusPacket = class ClientCacheBlobStatusPacket extends packet_1.Packet {
};
ClientCacheBlobStatusPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ClientCacheBlobStatusPacket);
exports.ClientCacheBlobStatusPacket = ClientCacheBlobStatusPacket;
let ClientCacheMissResponsePacket = class ClientCacheMissResponsePacket extends packet_1.Packet {
};
ClientCacheMissResponsePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ClientCacheMissResponsePacket);
exports.ClientCacheMissResponsePacket = ClientCacheMissResponsePacket;
let EducationSettingsPacket = class EducationSettingsPacket extends packet_1.Packet {
};
EducationSettingsPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], EducationSettingsPacket);
exports.EducationSettingsPacket = EducationSettingsPacket;
let EmotePacket = class EmotePacket extends packet_1.Packet {
};
EmotePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], EmotePacket);
exports.EmotePacket = EmotePacket;
let MultiplayerSettingsPacket = class MultiplayerSettingsPacket extends packet_1.Packet {
};
MultiplayerSettingsPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], MultiplayerSettingsPacket);
exports.MultiplayerSettingsPacket = MultiplayerSettingsPacket;
let SettingsCommandPacket = class SettingsCommandPacket extends packet_1.Packet {
};
SettingsCommandPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], SettingsCommandPacket);
exports.SettingsCommandPacket = SettingsCommandPacket;
let AnvilDamagePacket = class AnvilDamagePacket extends packet_1.Packet {
};
AnvilDamagePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], AnvilDamagePacket);
exports.AnvilDamagePacket = AnvilDamagePacket;
let CompletedUsingItemPacket = class CompletedUsingItemPacket extends packet_1.Packet {
};
CompletedUsingItemPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], CompletedUsingItemPacket);
exports.CompletedUsingItemPacket = CompletedUsingItemPacket;
let NetworkSettingsPacket = class NetworkSettingsPacket extends packet_1.Packet {
};
NetworkSettingsPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], NetworkSettingsPacket);
exports.NetworkSettingsPacket = NetworkSettingsPacket;
let PlayerAuthInputPacket = class PlayerAuthInputPacket extends packet_1.Packet {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(blockpos_1.Vec3)
], PlayerAuthInputPacket.prototype, "pos", void 0);
PlayerAuthInputPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], PlayerAuthInputPacket);
exports.PlayerAuthInputPacket = PlayerAuthInputPacket;
let CreativeContentPacket = class CreativeContentPacket extends packet_1.Packet {
};
CreativeContentPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], CreativeContentPacket);
exports.CreativeContentPacket = CreativeContentPacket;
let PlayerEnchantOptionsPacket = class PlayerEnchantOptionsPacket extends packet_1.Packet {
};
PlayerEnchantOptionsPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], PlayerEnchantOptionsPacket);
exports.PlayerEnchantOptionsPacket = PlayerEnchantOptionsPacket;
let ItemStackRequest = class ItemStackRequest extends packet_1.Packet {
};
ItemStackRequest = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ItemStackRequest);
exports.ItemStackRequest = ItemStackRequest;
let ItemStackResponse = class ItemStackResponse extends packet_1.Packet {
};
ItemStackResponse = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ItemStackResponse);
exports.ItemStackResponse = ItemStackResponse;
let PlayerArmorDamagePacket = class PlayerArmorDamagePacket extends packet_1.Packet {
};
PlayerArmorDamagePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], PlayerArmorDamagePacket);
exports.PlayerArmorDamagePacket = PlayerArmorDamagePacket;
let CodeBuilderPacket = class CodeBuilderPacket extends packet_1.Packet {
};
CodeBuilderPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], CodeBuilderPacket);
exports.CodeBuilderPacket = CodeBuilderPacket;
let UpdatePlayerGameTypePacket = class UpdatePlayerGameTypePacket extends packet_1.Packet {
};
UpdatePlayerGameTypePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], UpdatePlayerGameTypePacket);
exports.UpdatePlayerGameTypePacket = UpdatePlayerGameTypePacket;
let EmoteListPacket = class EmoteListPacket extends packet_1.Packet {
};
EmoteListPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], EmoteListPacket);
exports.EmoteListPacket = EmoteListPacket;
let PositionTrackingDBServerBroadcast = class PositionTrackingDBServerBroadcast extends packet_1.Packet {
};
PositionTrackingDBServerBroadcast = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], PositionTrackingDBServerBroadcast);
exports.PositionTrackingDBServerBroadcast = PositionTrackingDBServerBroadcast;
let PositionTrackingDBClientRequest = class PositionTrackingDBClientRequest extends packet_1.Packet {
};
PositionTrackingDBClientRequest = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], PositionTrackingDBClientRequest);
exports.PositionTrackingDBClientRequest = PositionTrackingDBClientRequest;
let DebugInfoPacket = class DebugInfoPacket extends packet_1.Packet {
};
DebugInfoPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], DebugInfoPacket);
exports.DebugInfoPacket = DebugInfoPacket;
let PacketViolationWarningPacket = class PacketViolationWarningPacket extends packet_1.Packet {
};
PacketViolationWarningPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], PacketViolationWarningPacket);
exports.PacketViolationWarningPacket = PacketViolationWarningPacket;
let MotionPredictionHintsPacket = class MotionPredictionHintsPacket extends packet_1.Packet {
};
MotionPredictionHintsPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], MotionPredictionHintsPacket);
exports.MotionPredictionHintsPacket = MotionPredictionHintsPacket;
let AnimateEntityPacket = class AnimateEntityPacket extends packet_1.Packet {
};
AnimateEntityPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], AnimateEntityPacket);
exports.AnimateEntityPacket = AnimateEntityPacket;
let CameraShakePacket = class CameraShakePacket extends packet_1.Packet {
};
CameraShakePacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], CameraShakePacket);
exports.CameraShakePacket = CameraShakePacket;
let PlayerFogPacket = class PlayerFogPacket extends packet_1.Packet {
};
PlayerFogPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], PlayerFogPacket);
exports.PlayerFogPacket = PlayerFogPacket;
let CorrectPlayerMovePredictionPacket = class CorrectPlayerMovePredictionPacket extends packet_1.Packet {
};
CorrectPlayerMovePredictionPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], CorrectPlayerMovePredictionPacket);
exports.CorrectPlayerMovePredictionPacket = CorrectPlayerMovePredictionPacket;
let ItemComponentPacket = class ItemComponentPacket extends packet_1.Packet {
};
ItemComponentPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], ItemComponentPacket);
exports.ItemComponentPacket = ItemComponentPacket;
let FilterTextPacket = class FilterTextPacket extends packet_1.Packet {
};
FilterTextPacket = tslib_1.__decorate([
    nativeclass_1.nativeClass(null)
], FilterTextPacket);
exports.FilterTextPacket = FilterTextPacket;
exports.PacketIdToType = {
    0x01: LoginPacket,
    0x02: PlayStatusPacket,
    0x03: ServerToClientHandshakePacket,
    0x04: ClientToServerHandshakePacket,
    0x05: DisconnectPacket,
    0x06: ResourcePacksInfoPacket,
    0x07: ResourcePacksStackPacket,
    0x08: ResourcePackClientResponsePacket,
    0x09: TextPacket,
    0x0a: SetTimePacket,
    0x0b: StartGamePacket,
    0x0c: AddPlayerPacket,
    0x0d: AddActorPacket,
    0x0e: RemoveActorPacket,
    0x0f: AddItemActorPacket,
    0x11: TakeItemActorPacket,
    0x12: MoveActorAbsolutePacket,
    0x13: MovePlayerPacket,
    0x14: RiderJumpPacket,
    0x15: UpdateBlockPacket,
    0x16: AddPaintingPacket,
    0x17: TickSyncPacket,
    0x18: LevelSoundEventPacketV1,
    0x19: LevelEventPacket,
    0x1a: BlockEventPacket,
    0x1b: ActorEventPacket,
    0x1c: MobEffectPacket,
    0x1d: UpdateAttributesPacket,
    0x1e: InventoryTransactionPacket,
    0x1f: MobEquipmentPacket,
    0x20: MobArmorEquipmentPacket,
    0x21: InteractPacket,
    0x22: BlockPickRequestPacket,
    0x23: ActorPickRequestPacket,
    0x24: PlayerActionPacket,
    0x26: HurtArmorPacket,
    0x27: SetActorDataPacket,
    0x28: SetActorMotionPacket,
    0x29: SetActorLinkPacket,
    0x2a: SetHealthPacket,
    0x2b: SetSpawnPositionPacket,
    0x2c: AnimatePacket,
    0x2d: RespawnPacket,
    0x2e: ContainerOpenPacket,
    0x2f: ContainerClosePacket,
    0x30: PlayerHotbarPacket,
    0x31: InventoryContentPacket,
    0x32: InventorySlotPacket,
    0x33: ContainerSetDataPacket,
    0x34: CraftingDataPacket,
    0x35: CraftingEventPacket,
    0x36: GuiDataPickItemPacket,
    0x37: AdventureSettingsPacket,
    0x38: BlockActorDataPacket,
    0x39: PlayerInputPacket,
    0x3a: LevelChunkPacket,
    0x3b: SetCommandsEnabledPacket,
    0x3c: SetDifficultyPacket,
    0x3d: ChangeDimensionPacket,
    0x3e: SetPlayerGameTypePacket,
    0x3f: PlayerListPacket,
    0x40: SimpleEventPacket,
    0x41: TelemetryEventPacket,
    0x42: SpawnExperienceOrbPacket,
    0x43: MapItemDataPacket,
    0x44: MapInfoRequestPacket,
    0x45: RequestChunkRadiusPacket,
    0x46: ChunkRadiusUpdatedPacket,
    0x47: ItemFrameDropItemPacket,
    0x48: GameRulesChangedPacket,
    0x49: CameraPacket,
    0x4a: BossEventPacket,
    0x4b: ShowCreditsPacket,
    0x4c: AvailableCommandsPacket,
    0x4d: CommandRequestPacket,
    0x4e: CommandBlockUpdatePacket,
    0x4f: CommandOutputPacket,
    0x52: ResourcePackDataInfoPacket,
    0x53: ResourcePackChunkDataPacket,
    0x54: ResourcePackChunkRequestPacket,
    0x55: TransferPacket,
    0x56: PlaySoundPacket,
    0x57: StopSoundPacket,
    0x58: SetTitlePacket,
    0x59: AddBehaviorTreePacket,
    0x5a: StructureBlockUpdatePacket,
    0x5b: ShowStoreOfferPacket,
    0x5c: PurchaseReceiptPacket,
    0x5d: PlayerSkinPacket,
    0x5e: SubClientLoginPacket,
    0x5f: WSConnectPacket,
    0x60: SetLastHurtByPacket,
    0x61: BookEditPacket,
    0x62: NpcRequestPacket,
    0x63: PhotoTransferPacket,
    0x64: ShowModalFormPacket,
    0x65: ModalFormResponsePacket,
    0x66: ServerSettingsRequestPacket,
    0x67: ServerSettingsResponsePacket,
    0x68: ShowProfilePacket,
    0x69: SetDefaultGameTypePacket,
    0x6a: RemoveObjectivePacket,
    0x6b: SetDisplayObjectivePacket,
    0x6c: SetScorePacket,
    0x6d: LabTablePacket,
    0x6e: UpdateBlockPacketSynced,
    0x6f: MoveActorDeltaPacket,
    0x70: SetScoreboardIdentityPacket,
    0x71: SetLocalPlayerAsInitializedPacket,
    0x72: UpdateSoftEnumPacket,
    0x73: NetworkStackLatencyPacket,
    0x75: ScriptCustomEventPacket,
    0x76: SpawnParticleEffect,
    0x77: AvailableActorIdentifiersPacket,
    0x78: LevelSoundEventPacketV2,
    0x79: NetworkChunkPublisherUpdatePacket,
    0x7a: BiomeDefinitionList,
    0x7b: LevelSoundEventPacket,
    0x7c: LevelEventGenericPacket,
    0x7d: LecternUpdatePacket,
    0x80: RemoveEntityPacket,
    0x81: ClientCacheStatusPacket,
    0x82: OnScreenTextureAnimationPacket,
    0x83: MapCreateLockedCopy,
    0x84: StructureTemplateDataRequestPacket,
    0x85: StructureTemplateDataExportPacket,
    0x87: ClientCacheBlobStatusPacket,
    0x88: ClientCacheMissResponsePacket,
    0x89: EducationSettingsPacket,
    0x8a: EmotePacket,
    0x8b: MultiplayerSettingsPacket,
    0x8c: SettingsCommandPacket,
    0x8d: AnvilDamagePacket,
    0x8e: CompletedUsingItemPacket,
    0x8f: NetworkSettingsPacket,
    0x90: PlayerAuthInputPacket,
    0x91: CreativeContentPacket,
    0x92: PlayerEnchantOptionsPacket,
    0x93: ItemStackRequest,
    0x94: ItemStackResponse,
    0x95: PlayerArmorDamagePacket,
    0x96: CodeBuilderPacket,
    0x97: UpdatePlayerGameTypePacket,
    0x98: EmoteListPacket,
    0x99: PositionTrackingDBServerBroadcast,
    0x9a: PositionTrackingDBClientRequest,
    0x9b: DebugInfoPacket,
    0x9c: PacketViolationWarningPacket,
    0x9d: MotionPredictionHintsPacket,
    0x9e: AnimateEntityPacket,
    0x9f: CameraShakePacket,
    0xa0: PlayerFogPacket,
    0xa1: CorrectPlayerMovePredictionPacket,
    0xa2: ItemComponentPacket,
    0xa3: FilterTextPacket,
};
for (const packetId in exports.PacketIdToType) {
    exports.PacketIdToType[packetId].ID = +packetId;
}
