// bdsx/bds/player.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerPlayer = exports.Player = void 0;
const common_1 = require("bdsx/common");
const pointer_1 = require("bdsx/pointer");
const actor_1 = require("./actor");
class Player extends actor_1.Actor {
    _setName(name) {
        common_1.abstract();
    }
    changeDimension(dimensionId, respawn) {
        common_1.abstract();
    }
    setName(name) {
        const _name = new pointer_1.CxxStringWrapper(true);
        _name.construct();
        _name.value = name;
        this._setName(_name);
        _name.destruct();
    }
    teleportTo(position, checkForBlocks, c, actorType, actorId) {
        common_1.abstract();
    }
    getInventory() {
        common_1.abstract();
    }
}
exports.Player = Player;
class ServerPlayer extends Player {
    _sendInventory() {
        common_1.abstract();
    }
    openInventory() {
        common_1.abstract();
    }
    sendNetworkPacket(packet) {
        common_1.abstract();
    }
    sendInventory() {
        setTimeout(() => {
            this._sendInventory();
        }, 50);
    }
}
exports.ServerPlayer = ServerPlayer;
