// bdsx/bds/raknet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RakNet = void 0;
const tslib_1 = require("tslib");
const common_1 = require("bdsx/common");
const nativetype_1 = require("bdsx/nativetype");
const nativeclass_1 = require("bdsx/nativeclass");
const core_1 = require("bdsx/core");
const portDelineator = '|'.charCodeAt(0);
var RakNet;
(function (RakNet) {
    let SystemAddress = class SystemAddress extends nativeclass_1.NativeClass {
        // void SystemAddress::ToString(bool writePort, char *dest, char portDelineator) const
        ToString(writePort, dest, portDelineator) {
            common_1.abstract();
        }
        toString() {
            const dest = Buffer.alloc(128);
            this.ToString(true, dest, portDelineator);
            const len = dest.indexOf(0);
            if (len === -1)
                throw Error('SystemAddress.ToString failed, null character not found');
            return dest.subarray(0, len).toString();
        }
    };
    tslib_1.__decorate([
        nativeclass_1.nativeField(nativetype_1.uint16_t, 130)
    ], SystemAddress.prototype, "systemIndex", void 0);
    SystemAddress = tslib_1.__decorate([
        nativeclass_1.nativeClass(136)
    ], SystemAddress);
    RakNet.SystemAddress = SystemAddress;
    let RakNetGUID = class RakNetGUID extends nativeclass_1.NativeClass {
    };
    tslib_1.__decorate([
        nativeclass_1.nativeField(nativetype_1.bin64_t)
    ], RakNetGUID.prototype, "g", void 0);
    tslib_1.__decorate([
        nativeclass_1.nativeField(nativetype_1.uint16_t)
    ], RakNetGUID.prototype, "systemIndex", void 0);
    RakNetGUID = tslib_1.__decorate([
        nativeclass_1.nativeClass()
    ], RakNetGUID);
    RakNet.RakNetGUID = RakNetGUID;
    let RakPeer = class RakPeer extends nativeclass_1.NativeClass {
        GetSystemAddressFromIndex(idx) {
            common_1.abstract();
        }
    };
    tslib_1.__decorate([
        nativeclass_1.nativeField(core_1.VoidPointer)
    ], RakPeer.prototype, "vftable", void 0);
    RakPeer = tslib_1.__decorate([
        nativeclass_1.nativeClass()
    ], RakPeer);
    RakNet.RakPeer = RakPeer;
    RakNet.UNASSIGNED_RAKNET_GUID = new RakNetGUID(true);
    RakNet.UNASSIGNED_RAKNET_GUID.g = nativetype_1.bin64_t.minus_one;
    RakNet.UNASSIGNED_RAKNET_GUID.systemIndex = -1;
    let AddressOrGUID = class AddressOrGUID extends nativeclass_1.NativeClass {
        GetSystemIndex() {
            const rakNetGuid = this.rakNetGuid;
            if (rakNetGuid !== RakNet.UNASSIGNED_RAKNET_GUID) {
                return rakNetGuid.systemIndex;
            }
            else {
                return this.systemAddress.systemIndex;
            }
        }
    };
    tslib_1.__decorate([
        nativeclass_1.nativeField(RakNetGUID)
    ], AddressOrGUID.prototype, "rakNetGuid", void 0);
    tslib_1.__decorate([
        nativeclass_1.nativeField(SystemAddress)
    ], AddressOrGUID.prototype, "systemAddress", void 0);
    AddressOrGUID = tslib_1.__decorate([
        nativeclass_1.nativeClass()
    ], AddressOrGUID);
    RakNet.AddressOrGUID = AddressOrGUID;
})(RakNet = exports.RakNet || (exports.RakNet = {}));
