// bdsx/bds/server.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serverInstance = exports.ServerInstance = exports.MinecraftServerScriptEngine = exports.ScriptFramework = exports.DedicatedServer = exports.Minecraft = exports.VanilaGameModuleServer = exports.Minecraft$Something = exports.EntityRegistryOwned = exports.VanilaServerGameplayEventListener = exports.ServerMetricsImpl = exports.ServerMetrics = exports.PrivateKeyManager = exports.Whitelist = exports.ResourcePackManager = exports.MinecraftEventing = void 0;
const nativeclass_1 = require("bdsx/nativeclass");
const common_1 = require("../common");
class MinecraftEventing extends nativeclass_1.NativeClass {
}
exports.MinecraftEventing = MinecraftEventing;
class ResourcePackManager extends nativeclass_1.NativeClass {
}
exports.ResourcePackManager = ResourcePackManager;
class Whitelist extends nativeclass_1.NativeClass {
}
exports.Whitelist = Whitelist;
class PrivateKeyManager extends nativeclass_1.NativeClass {
}
exports.PrivateKeyManager = PrivateKeyManager;
class ServerMetrics extends nativeclass_1.NativeClass {
}
exports.ServerMetrics = ServerMetrics;
class ServerMetricsImpl extends ServerMetrics {
}
exports.ServerMetricsImpl = ServerMetricsImpl;
class VanilaServerGameplayEventListener extends nativeclass_1.NativeClass {
}
exports.VanilaServerGameplayEventListener = VanilaServerGameplayEventListener;
class EntityRegistryOwned extends nativeclass_1.NativeClass {
}
exports.EntityRegistryOwned = EntityRegistryOwned;
/**
 * unknown instance
 */
class Minecraft$Something extends nativeclass_1.NativeClass {
}
exports.Minecraft$Something = Minecraft$Something;
class VanilaGameModuleServer extends nativeclass_1.NativeClass {
}
exports.VanilaGameModuleServer = VanilaGameModuleServer;
class Minecraft extends nativeclass_1.NativeClass {
    getLevel() {
        common_1.abstract();
    }
}
exports.Minecraft = Minecraft;
class DedicatedServer extends nativeclass_1.NativeClass {
}
exports.DedicatedServer = DedicatedServer;
// struct DedicatedServer :ScriptCommandOrigin
// {
//     struct Something
//     {
//         Vector<VoidPointer> vector1;
//         Vector<VoidPointer> vector2;
//         VoidPointer u1;
//         VoidPointer u2;
//     };
//     DedicatedServer() = delete;
class ScriptFramework extends nativeclass_1.NativeClass {
}
exports.ScriptFramework = ScriptFramework;
class MinecraftServerScriptEngine extends ScriptFramework {
}
exports.MinecraftServerScriptEngine = MinecraftServerScriptEngine;
class ServerInstance extends nativeclass_1.NativeClass {
    createDimension(id) {
        return this.minecraft.something.level.createDimension(id);
    }
    getActivePlayerCount() {
        return this.minecraft.something.level.getActivePlayerCount();
    }
    disconnectClient(client, message = "disconnectionScreen.disconnected") {
        return this.minecraft.something.shandler.disconnectClient(client, message);
    }
    setMotd(motd) {
        return this.minecraft.something.shandler.setMotd(motd);
    }
}
exports.ServerInstance = ServerInstance;
