// bdsx/eminus/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.eminus_load_dlls_in_mods = void 0;
const core_1 = require("bdsx/core");
const dll_1 = require("bdsx/dll");
const util_1 = require("bdsx/util");
const windows_h_1 = require("bdsx/windows_h");
const dllchecker_1 = require("./dllchecker");
const fs = require("fs");
const colors = require("colors");
const common_1 = require("bdsx/common");
function getErrorMessage(errcode) {
    const bufferptr = new core_1.AllocatedPointer(8);
    const charlen = dll_1.dll.kernel32.FormatMessageW(windows_h_1.FORMAT_MESSAGE_ALLOCATE_BUFFER | windows_h_1.FORMAT_MESSAGE_FROM_SYSTEM, null, errcode, windows_h_1.MAKELANGID(windows_h_1.LANG_NEUTRAL, windows_h_1.SUBLANG_DEFAULT), bufferptr, 0, null);
    const strptr = bufferptr.getPointer();
    const message = strptr.getString(charlen, 0, common_1.Encoding.Utf16);
    dll_1.dll.kernel32.LocalFree(strptr);
    return message.trim();
}
function printErrorCode(err) {
    console.error(colors.red(`Error Code: ${err}`));
    console.error(colors.red('Error Message: ' + getErrorMessage(err)));
}
function loadEminus() {
    try {
        return fs.readFileSync('mods/eminus.ini', 'utf-8');
    }
    catch (err) {
        return null;
    }
}
var IniSection;
(function (IniSection) {
    IniSection[IniSection["Global"] = 0] = "Global";
    IniSection[IniSection["Module"] = 1] = "Module";
})(IniSection || (IniSection = {}));
/**
 * @deprecated use NativeModule.load
 */
function eminus_load_dlls_in_mods() {
    const modules = [];
    let verbose = true;
    const ini = loadEminus();
    if (ini !== null) {
        let section = IniSection.Global;
        let linenum = 0;
        for (let line of ini.split('\n')) {
            line = line.split('#', 1)[0].trim();
            linenum++;
            if (line === '') {
                // empty
                continue;
            }
            // section
            if (line[0] === '[') {
                let offset = 0;
                offset++;
                let sectionend = line.indexOf(']', offset);
                if (sectionend !== -1) {
                    line = line.substr(0, sectionend);
                    sectionend -= offset;
                }
                else {
                    sectionend = line.length - offset;
                }
                const sectionName = line.substring(offset, sectionend);
                switch (sectionName) {
                    case 'module':
                        section = IniSection.Module;
                        break;
                    case 'global':
                        section = IniSection.Global;
                        break;
                    default:
                        console.error(`[EMinus] eminus.ini(${linenum}): - unknown section ${line.substr(offset, sectionend)}`);
                        break;
                }
                continue;
            }
            // find equal
            const equal = line.indexOf('=');
            if (equal !== -1) {
                const name = line.substr(0, equal).trimRight();
                const value = line.substr(equal + 1).trimLeft();
                switch (section) {
                    case IniSection.Global:
                        if (name === 'verbose') {
                            verbose = !!+value;
                            continue;
                        }
                        break;
                }
                console.error(`"[EMinus] eminus.ini(${linenum}): - unknown property ${name}`);
            }
            else {
                // name only
                switch (section) {
                    case IniSection.Module:
                        modules.push(line);
                        break;
                    default:
                        console.error(`[EMinus] eminus.ini(${linenum}): - unknown property ${line}`);
                        break;
                }
            }
        }
    }
    else {
        if (!util_1.isDirectory('mods')) {
            console.error('[EMinus] no mods directory, skip');
            return;
        }
        console.error('[EMinus] no mods\\eminus.ini, It will load mods\\*.dll');
        modules.push(...fs.readdirSync('./mods'));
    }
    console.log(colors.red('[EMinus] Element Minus is deprecated. Please use NativeModule.load directly'));
    if (verbose)
        console.log('[EMinus] JS-Version');
    dll_1.dll.kernel32.SetDllDirectoryW('mods');
    for (const name of modules) {
        try {
            dll_1.NativeModule.get(name);
            if (verbose)
                console.log(`[EMinus] mods\\${name} (Already loaded)`);
            continue;
        }
        catch (err) {
        }
        if (verbose)
            console.log(`[EMinus] mods\\${name}`);
        try {
            dll_1.NativeModule.load(name);
            continue;
        }
        catch (err) {
            console.error(`[EMinus] mods\\${name}: ${colors.red('Failed')}`);
            if (err.errno) {
                printErrorCode(err.errno);
                if (err.errno === windows_h_1.ERROR_MOD_NOT_FOUND) {
                    dllchecker_1.dllchecker.check(name);
                }
                process.exit(windows_h_1.ERROR_MOD_NOT_FOUND);
            }
            else {
                throw err;
            }
        }
    }
}
exports.eminus_load_dlls_in_mods = eminus_load_dlls_in_mods;
