// bdsx/chat.js
"use strict";
const krevent_1 = require("krevent");
const packetids_1 = require("./bds/packetids");
const common_1 = require("./common");
const nethook_1 = require("./nethook");
class ChatEventImpl {
    constructor(name, message, networkIdentifier) {
        this.name = name;
        this.message = message;
        this.networkIdentifier = networkIdentifier;
        this.isModified = false;
    }
    setName(name) {
        this.isModified = true;
        this.name = name;
    }
    setMessage(message) {
        this.isModified = true;
        this.message = message;
    }
}
class ChatManager extends krevent_1.EventEx {
    constructor() {
        super(...arguments);
        this.chatlistener = (ptr, networkIdentifier, packetId) => {
            const name = ptr.name;
            const message = ptr.message;
            const ev = new ChatEventImpl(name, message, networkIdentifier);
            if (this.fire(ev) === common_1.CANCEL)
                return common_1.CANCEL;
            if (ev.isModified) {
                ptr.name = ev.name;
                ptr.message = ev.message;
            }
        };
    }
    onStarted() {
        nethook_1.nethook.before(packetids_1.MinecraftPacketIds.Text).on(this.chatlistener);
    }
    onCleared() {
        nethook_1.nethook.before(packetids_1.MinecraftPacketIds.Text).remove(this.chatlistener);
    }
    /** @deprecated use nethook.before(MinecraftPacketIds.Text).on */
    on(listener) {
        super.on(listener);
    }
}
module.exports = new ChatManager();
