// bdsx/command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hook = exports.command = exports.net = exports.hookingForCommand = void 0;
const krevent_1 = require("krevent");
const command_1 = require("./bds/command");
const packetids_1 = require("./bds/packetids");
const proc_1 = require("./bds/proc");
const common_1 = require("./common");
const makefunc_1 = require("./makefunc");
const nethook_1 = require("./nethook");
const sharedpointer_1 = require("./sharedpointer");
const util_1 = require("./util");
function hookingForCommand() {
    const executeCommandOriginal = proc_1.procHacker.hooking('MinecraftCommands::executeCommand', command_1.MCRESULT, null, command_1.MinecraftCommands, command_1.MCRESULT, sharedpointer_1.SharedPtr.make(command_1.CommandContext), makefunc_1.RawTypeId.Boolean)((cmd, res, ctxptr, b) => {
        const ctx = ctxptr.p;
        const name = ctx.origin.getName();
        const resv = hookev.fire(ctxptr.p.command, name, ctx);
        switch (typeof resv) {
            case 'number':
                res.result = resv;
                util_1._tickCallback();
                return res;
            default:
                util_1._tickCallback();
                return executeCommandOriginal(cmd, res, ctxptr, b);
        }
    });
}
exports.hookingForCommand = hookingForCommand;
class CommandEventImpl {
    constructor(command, networkIdentifier) {
        this.command = command;
        this.networkIdentifier = networkIdentifier;
        this.isModified = false;
    }
    setCommand(command) {
        this.isModified = true;
        this.command = command;
    }
}
class UserCommandEvents extends krevent_1.EventEx {
    constructor() {
        super(...arguments);
        this.listener = (ptr, networkIdentifier, packetId) => {
            const command = ptr.command;
            const ev = new CommandEventImpl(command, networkIdentifier);
            if (this.fire(ev) === common_1.CANCEL)
                return common_1.CANCEL;
            if (ev.isModified) {
                ptr.command = ev.command;
            }
        };
    }
    onStarted() {
        nethook_1.nethook.before(packetids_1.MinecraftPacketIds.CommandRequest).on(this.listener);
    }
    onCleared() {
        nethook_1.nethook.before(packetids_1.MinecraftPacketIds.CommandRequest).remove(this.listener);
    }
}
const hookev = new krevent_1.default();
/** @deprecated use nethook.before(MinecraftPacketIds.CommandRequest).on */
exports.net = new UserCommandEvents();
var command;
(function (command) {
    command.hook = hookev;
})(command = exports.command || (exports.command = {}));
/**
 * @deprecated use command.hook
 */
exports.hook = hookev;
