// bdsx/commandlist.js
"use strict";
/**
 * on progressing
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const bdsx_1 = require("bdsx");
const command_1 = require("bdsx/bds/command");
const server_1 = require("bdsx/bds/server");
const makefunc_1 = require("bdsx/makefunc");
const nativeclass_1 = require("bdsx/nativeclass");
const nativetype_1 = require("bdsx/nativetype");
class Command extends nativeclass_1.NativeClass {
    static makeAllocator() {
        const cls = this;
        const size = cls[nativetype_1.NativeType.size];
        if (!size)
            throw Error(`${cls.name}: size is not defined`);
        return makefunc_1.makefunc.np((returnval) => {
            const ptr = bdsx_1.capi.malloc(size);
            const cmd = ptr.as(cls);
            returnval.setPointer(cmd);
            return returnval;
        }, bdsx_1.StaticPointer, null, bdsx_1.StaticPointer);
    }
}
(function (Command) {
    class VFTable extends nativeclass_1.NativeClass {
    }
    Command.VFTable = VFTable;
})(Command || (Command = {}));
Command.define({
    vftable: Command.VFTable,
    u1: bdsx_1.VoidPointer,
}, 0x20);
class CustomCommand extends Command {
}
const COL_SIG_REV0 = 0;
const COL_SIG_REV1 = 1;
let _s_RTTICompleteObjectLocator = class _s_RTTICompleteObjectLocator extends nativeclass_1.NativeClass {
};
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.int32_t)
], _s_RTTICompleteObjectLocator.prototype, "signature", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.int32_t)
], _s_RTTICompleteObjectLocator.prototype, "offset", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.int32_t)
], _s_RTTICompleteObjectLocator.prototype, "cdOffset", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.int32_t)
], _s_RTTICompleteObjectLocator.prototype, "pTypeDescriptor", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.int32_t)
], _s_RTTICompleteObjectLocator.prototype, "pClassDescriptor", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.int32_t)
], _s_RTTICompleteObjectLocator.prototype, "pSelf", void 0);
_s_RTTICompleteObjectLocator = tslib_1.__decorate([
    nativeclass_1.nativeClass()
], _s_RTTICompleteObjectLocator);
// import './example_and_test'; // remove this if it's not necessary for you
// commands.list.description
const cmdreg = server_1.serverInstance.minecraft.commands.registry;
cmdreg.registerCommand('testcommand', 'test command', command_1.CommandPermissionLevel.Normal, 8, 40);
const cmd = cmdreg.findCommand('testcommand');
const overload = new command_1.CommandRegistry.Overload(true);
overload.commandVersion = bdsx_1.bin.make64(1, 0x7fffffff);
overload.allocator = CustomCommand.makeAllocator();
overload.u3 = nativetype_1.bin64_t.zero;
overload.u4 = nativetype_1.bin64_t.zero;
overload.u5 = nativetype_1.bin64_t.zero;
overload.u6 = -1;
cmd.overloads.push(overload);
cmdreg.registerOverloadInternal(cmd, overload);
