// bdsx/event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.events = void 0;
const bdsx_1 = require("bdsx");
const block_1 = require("bdsx/bds/block");
const blockpos_1 = require("bdsx/bds/blockpos");
const gamemode_1 = require("bdsx/bds/gamemode");
const inventory_1 = require("bdsx/bds/inventory");
const player_1 = require("bdsx/bds/player");
const proc_1 = require("bdsx/bds/proc");
const common_1 = require("bdsx/common");
const core_1 = require("bdsx/core");
const krevent_1 = require("krevent");
class BlockDestroyEvent {
    constructor(player, blockPos) {
        this.player = player;
        this.blockPos = blockPos;
    }
}
function onBlockDestroy(survivalMode, blockPos, v) {
    const event = new BlockDestroyEvent(survivalMode.actor, blockPos);
    if (exports.events.blockDestroy.fire(event) === common_1.CANCEL) {
        return false;
    }
    else {
        return _onBlockDestroy(survivalMode, blockPos, v);
    }
}
function onBlockDestroyCreative(gameMode, blockPos, v) {
    const event = new BlockDestroyEvent(gameMode.actor, blockPos);
    if (exports.events.blockDestroy.fire(event) === common_1.CANCEL) {
        return false;
    }
    else {
        return _onBlockDestroyCreative(gameMode, blockPos, v);
    }
}
const _onBlockDestroy = proc_1.procHacker.hooking("SurvivalMode::destroyBlock", bdsx_1.RawTypeId.Boolean, null, gamemode_1.SurvivalMode, blockpos_1.BlockPos, bdsx_1.RawTypeId.Int32)(onBlockDestroy);
const _onBlockDestroyCreative = proc_1.procHacker.hooking("GameMode::_creativeDestroyBlock", bdsx_1.RawTypeId.Boolean, null, gamemode_1.SurvivalMode, blockpos_1.BlockPos, bdsx_1.RawTypeId.Int32)(onBlockDestroyCreative);
class BlockPlaceEvent {
    constructor(player, block, blockSource, blockPos) {
        this.player = player;
        this.block = block;
        this.blockSource = blockSource;
        this.blockPos = blockPos;
    }
}
function onBlockPlace(blockSource, block, blockPos, v1, actor, v2) {
    const event = new BlockPlaceEvent(actor, block, blockSource, blockPos);
    if (exports.events.blockPlace.fire(event) === common_1.CANCEL) {
        return false;
    }
    else {
        return _onBlockPlace(blockSource, block, blockPos, v1, actor, v2);
    }
}
const _onBlockPlace = proc_1.procHacker.hooking("BlockSource::mayPlace", bdsx_1.RawTypeId.Boolean, null, block_1.BlockSource, block_1.Block, blockpos_1.BlockPos, bdsx_1.RawTypeId.Int32, bdsx_1.Actor, bdsx_1.RawTypeId.Boolean)(onBlockPlace);
class PlayerAttackEvent {
    constructor(player, victim) {
        this.player = player;
        this.victim = victim;
    }
}
function onPlayerAttack(player, victim) {
    const event = new PlayerAttackEvent(player, victim);
    if (exports.events.playerAttack.fire(event) === common_1.CANCEL) {
        return false;
    }
    else {
        return _onPlayerAttack(player, victim);
    }
}
const _onPlayerAttack = proc_1.procHacker.hooking("Player::attack", bdsx_1.RawTypeId.Boolean, null, player_1.Player, bdsx_1.Actor)(onPlayerAttack);
class PlayerDropItemEvent {
    constructor(player, itemStack) {
        this.player = player;
        this.itemStack = itemStack;
    }
}
function onPlayerDropItem(player, itemStack, v) {
    const event = new PlayerDropItemEvent(player, itemStack);
    if (exports.events.playerDropItem.fire(event) === common_1.CANCEL) {
        return false;
    }
    else {
        return _onPlayerDropItem(player, itemStack, v);
    }
}
const _onPlayerDropItem = proc_1.procHacker.hooking("Player::drop", bdsx_1.RawTypeId.Boolean, null, player_1.Player, inventory_1.ItemStack, bdsx_1.RawTypeId.Boolean)(onPlayerDropItem);
class PlayerJoinEvent {
    constructor(player) {
        this.player = player;
    }
}
bdsx_1.nethook.send(bdsx_1.MinecraftPacketIds.PlayStatus).on((pk, ni) => {
    if (pk.status === 3) {
        const event = new PlayerJoinEvent(ni.getActor());
        exports.events.playerJoin.fire(event);
    }
});
class PlayerPickupItemEvent {
    constructor(player) {
        this.player = player;
    }
}
function onPlayerPickupItem(player, itemActor, v1, v2) {
    const event = new PlayerPickupItemEvent(player /*, itemActor.itemStack*/);
    if (exports.events.playerPickupItem.fire(event) === common_1.CANCEL) {
        return false;
    }
    else {
        return _onPlayerPickupItem(player, itemActor, v1, v2);
    }
}
const _onPlayerPickupItem = proc_1.procHacker.hooking("Player::take", bdsx_1.RawTypeId.Boolean, null, player_1.Player, core_1.VoidPointer, bdsx_1.RawTypeId.Int32, bdsx_1.RawTypeId.Int32)(onPlayerPickupItem);
exports.events = {
    /** Cancellable */
    blockDestroy: new krevent_1.default(),
    /** Cancellable */
    blockPlace: new krevent_1.default(),
    /** Cancellable */
    playerAttack: new krevent_1.default(),
    /** Cancellable but only when player is in container screens*/
    playerDropItem: new krevent_1.default(),
    /** Not cancellable */
    playerJoin: new krevent_1.default(),
    /** Cancellable */
    playerPickupItem: new krevent_1.default(),
};
