// bdsx/native.js
"use strict";
/**
 * @deprecated redirection for old bdsx
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadPdb = exports.free = exports.malloc = exports.std$_Allocate$16 = exports.NativeModule = exports.SharedPointer = exports.Actor = exports.NativePointer = exports.StaticPointer = exports.NetworkIdentifier = exports.nethook = exports.serverControl = exports.ipfilter = exports.setOnCommandListener = exports.setOnRuntimeErrorListener = exports.setOnErrorListener = void 0;
const core = require("./core");
const actor_1 = require("./bds/actor");
const networkidentifier_1 = require("./bds/networkidentifier");
const capi_1 = require("./capi");
const cxxvector_1 = require("./cxxvector");
const dll_1 = require("./dll");
const legacy_1 = require("./legacy");
const nethook_1 = require("./nethook");
const servercontrol_1 = require("./servercontrol");
const sharedpointer_1 = require("./sharedpointer");
function defaultErrorMethod(err) {
    console.error(err);
}
/**
 * Catch global errors
 * default error printing is disabled if cb returns false
 * @deprecated use bedrockServer.error.on
 */
function setOnErrorListener(cb) {
    if (cb) {
        core.jshook.setOnError(err => {
            if (cb(err) !== false) {
                console.error(err.stack || err.toString());
            }
        });
    }
    else {
        core.jshook.setOnError(defaultErrorMethod);
    }
}
exports.setOnErrorListener = setOnErrorListener;
/**
 * @deprecated use bedrockServer.close.on
 */
exports.setOnRuntimeErrorListener = legacy_1.legacy.setOnRuntimeErrorListener;
/**
 * @deprecated
 */
function setOnCommandListener(cb) {
    throw Error('not implemented');
}
exports.setOnCommandListener = setOnCommandListener;
/**
 * @deprecated use bdsx.ipfilter
 */
exports.ipfilter = core.ipfilter;
/**
 * @deprecated use bdsx.serverControl
 */
exports.serverControl = servercontrol_1.serverControl;
/**
 * @deprecated use bdsx.nethook
*/
exports.nethook = nethook_1.nethook;
/**
 * @deprecated use bdsx.NetworkIdentifier
*/
exports.NetworkIdentifier = networkidentifier_1.NetworkIdentifier;
/**
 * @deprecated use bdsx.StaticPointer
*/
exports.StaticPointer = core.StaticPointer;
/**
 * @deprecated use bdsx.NativePointer
*/
exports.NativePointer = core.NativePointer;
/**
 * @deprecated use bdsx.Actor
*/
exports.Actor = actor_1.Actor;
/**
 * @deprecated use bdsx.SharedPtr
 */
exports.SharedPointer = sharedpointer_1.SharedPointer;
/**
 * @deprecated use bdsx.NativeModule
 */
exports.NativeModule = dll_1.NativeModule;
/**
 * the alloc function for std::vector
 * @deprecated why are you using it?, use capi.malloc
 */
exports.std$_Allocate$16 = cxxvector_1.CxxVector._alloc16;
/**
 * memory allocate by native c
 * @deprecated use capi.malloc
 */
exports.malloc = capi_1.capi.malloc;
/**
 * memory release by native c
 * @deprecated use capi.free
 */
exports.free = capi_1.capi.free;
/**
 * @deprecated use analyzer.loadMap
 */
function loadPdb() {
    return core.pdb.getAll();
}
exports.loadPdb = loadPdb;
