// bdsx/plugins.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadAllPlugins = void 0;
const fs_1 = require("fs");
const path = require("path");
const colors = require("colors");
const source_map_support_1 = require("./source-map-support");
const concurrency_1 = require("./concurrency");
class PromCounter {
    constructor() {
        this.resolve = null;
        this.counter = 0;
        this.prom = Promise.resolve();
    }
    ref() {
        if (this.counter === 0) {
            this.prom = new Promise(resolve => {
                this.resolve = resolve;
            });
        }
        this.counter++;
    }
    unref() {
        this.counter--;
        if (this.counter === 0) {
            this.resolve();
            this.resolve = null;
        }
    }
    wait() {
        return this.prom;
    }
}
async function loadAllPlugins() {
    const taskQueue = new concurrency_1.ConcurrencyQueue;
    const loaded = new Set();
    async function loadPackageJson(name, json) {
        if (!json) {
            console.error(`[BDSX-Plugins] Invalid ${name}/package.json`);
            return false;
        }
        if (!json.dependencies)
            return true;
        const counter = new PromCounter;
        for (const name in json.dependencies) {
            if (!name.startsWith('@bdsx/'))
                continue;
            if (loaded.has(name))
                continue;
            loaded.add(name);
            counter.ref();
            taskQueue.run(async () => {
                try {
                    const jsonpath = require.resolve(name + '/package.json');
                    const json = JSON.parse(await fs_1.promises.readFile(jsonpath, 'utf-8'));
                    if (json.bdsxPlugin) {
                        await loadPackageJson(name, json);
                    }
                }
                catch (err) {
                    console.error(`[BDSX-Plugins] Failed to get ${name}/package.json`);
                }
                counter.unref();
            });
        }
        await counter.wait();
        return true;
    }
    try {
        const packagejson = path.resolve(process.cwd(), process.argv[1], 'package.json');
        const mainjson = JSON.parse(await fs_1.promises.readFile(packagejson, 'utf-8'));
        loadPackageJson('[entrypoint]', mainjson);
        await taskQueue.onceEnd();
        if (loaded.size === 0) {
            console.log('[BDSX-Plugins] No Plugins');
        }
        else {
            let index = 0;
            for (const name of loaded) {
                try {
                    console.log(`[BDSX-Plugins] (${++index}/${loaded.size}) ${name}`);
                    require(name);
                }
                catch (err) {
                    console.error(`[BDSX-Plugins] Failed to load ${name}`);
                    console.error(source_map_support_1.remapStack(err.stack));
                }
            }
        }
    }
    catch (err) {
        console.error(colors.red(`[BDSX-Plugins] Failed to load`));
        if (err.code === 'ENOENT') {
            console.error(colors.red(`[BDSX-Plugins] package.json not found. please set the entry point as package.json directory`));
        }
        else {
            console.error(colors.red(`[BDSX-Plugins] Failed to read package.json. ` + err.message));
        }
        return;
    }
}
exports.loadAllPlugins = loadAllPlugins;
