// bdsx/pointer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CxxStringPointer = exports.CxxStringStructure = exports.CxxStringWrapper = exports.Pointer = exports.Wrapper = void 0;
const common_1 = require("./common");
const dll_1 = require("./dll");
const nativeclass_1 = require("./nativeclass");
const nativetype_1 = require("./nativetype");
class Wrapper extends nativeclass_1.NativeClass {
    static make(type) {
        class TypedWrapper extends Wrapper {
        }
        TypedWrapper.prototype.type = type;
        TypedWrapper.define({ value: type });
        return TypedWrapper;
    }
    static [nativetype_1.NativeType.ctor_copy](to, from) {
        to.copyFrom(from, 8);
    }
    static [nativetype_1.NativeType.ctor_move](to, from) {
        to.copyFrom(from, 8);
    }
    static [nativetype_1.NativeType.descriptor](builder, key, offset) {
        const type = this;
        let obj = null;
        function init(ptr) {
            obj = ptr.getPointerAs(type, offset);
            Object.defineProperty(ptr, key, {
                get() {
                    return obj;
                },
                set(v) {
                    obj = v;
                    ptr.setPointer(v, offset);
                }
            });
        }
        builder.desc[key] = {
            configurable: true,
            get() {
                init(this);
                return obj;
            }
        };
    }
}
exports.Wrapper = Wrapper;
/** @deprecated renamed to Wrapper<T> */
class Pointer extends Wrapper {
    static make(type) {
        class TypedPointer extends Pointer {
        }
        TypedPointer.prototype.type = type;
        TypedPointer.defineAsUnion({ p: type, value: type });
        return TypedPointer;
    }
}
exports.Pointer = Pointer;
class CxxStringWrapper extends nativeclass_1.NativeClass {
    [nativetype_1.NativeType.ctor]() {
        common_1.abstract();
    }
    [nativetype_1.NativeType.dtor]() {
        common_1.abstract();
    }
    /**
     * @deprecated use .destruct
     */
    dispose() {
        this.destruct();
    }
    get value() {
        return this.getCxxString();
    }
    set value(str) {
        this.setCxxString(str);
    }
    get valueptr() {
        if (this.capacity >= 0x10)
            return this.getPointer();
        else
            return this.add();
    }
    reserve(nsize) {
        const capacity = this.capacity;
        if (nsize > capacity) {
            const orivalue = this.valueptr;
            this.capacity = nsize;
            const dest = dll_1.dll.ucrtbase.malloc(nsize + 1);
            dest.copyFrom(orivalue, this.length);
            if (capacity >= 0x10)
                dll_1.dll.ucrtbase.free(orivalue);
            this.setPointer(dest);
            if (dest === null) {
                this.setString("[out of memory]\0");
                this.capacity = 15;
                this.length = 15;
                return;
            }
        }
    }
    resize(nsize) {
        this.reserve(nsize);
        this.length = nsize;
    }
}
exports.CxxStringWrapper = CxxStringWrapper;
CxxStringWrapper.define({
    length: [nativetype_1.int64_as_float_t, 0x10],
    capacity: [nativetype_1.int64_as_float_t, 0x18]
});
const strctor = nativetype_1.CxxString[nativetype_1.NativeType.ctor];
const strdtor = nativetype_1.CxxString[nativetype_1.NativeType.dtor];
CxxStringWrapper.prototype[nativetype_1.NativeType.ctor] = function () { return strctor(this); };
CxxStringWrapper.prototype[nativetype_1.NativeType.dtor] = function () { return strdtor(this); };
/** @deprecated renamed to CxxStringWrapper */
exports.CxxStringStructure = CxxStringWrapper;
/** @deprecated use CxxStringWrapper */
exports.CxxStringPointer = Wrapper.make(nativetype_1.CxxString);
