// bdsx/prochacker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProcHacker = void 0;
const assembler_1 = require("./assembler");
const core_1 = require("./core");
const disassembler_1 = require("./disassembler");
const dll_1 = require("./dll");
const hacktool_1 = require("./hacktool");
const makefunc_1 = require("./makefunc");
const unlocker_1 = require("./unlocker");
const util_1 = require("./util");
const colors = require("colors");
const FREE_REGS = [
    assembler_1.Register.rax,
    assembler_1.Register.r10,
    assembler_1.Register.r11,
];
class AsmMover extends assembler_1.X64Assembler {
    constructor(origin, codesize) {
        super(new Uint8Array(32), 0);
        this.origin = origin;
        this.codesize = codesize;
        this.freeregs = new Set(FREE_REGS);
        this.inpos = 0;
    }
    getUnusing() {
        for (const r of this.freeregs.values()) {
            return r;
        }
        return null;
    }
    asmFromOrigin(oper) {
        const splits = oper.splits;
        const basename = splits[0];
        let ripDependedParam = null;
        const params = oper.parameters();
        for (const info of params) {
            switch (info.type) {
                case 'r':
                    this.freeregs.delete(info.register);
                    break;
                case 'rp':
                    this.freeregs.delete(info.register);
                    if (info.register === assembler_1.Register.rip) {
                        ripDependedParam = info;
                    }
                    break;
            }
        }
        if ((basename.startsWith('j') || basename === 'call') && splits.length === 2 && splits[1] === 'c') {
            // jump
            const offset = this.inpos + oper.args[0] + oper.size;
            if (offset < 0 || offset > this.codesize) {
                const tmpreg = this.getUnusing();
                if (tmpreg === null)
                    throw Error(`Not enough free registers`);
                const jmp_r = assembler_1.asm.code[basename + '_r'];
                if (jmp_r) {
                    this.mov_r_c(tmpreg, this.origin.add(offset));
                    jmp_r.call(this, tmpreg);
                }
                else {
                    const reversed = oper.reverseJump();
                    this[reversed + '_label']('!');
                    this.jmp64(this.origin.add(offset), tmpreg);
                    this.close_label('!');
                }
                this.inpos += oper.size;
                return;
            }
            else {
                // TOFIX: remap offset if the code size is changed when rewriting.
            }
        }
        else {
            if (ripDependedParam !== null) {
                const tmpreg = this.getUnusing();
                if (tmpreg === null)
                    throw Error(`Not enough free registers`);
                oper.args[ripDependedParam.argi] = tmpreg;
                oper.args[ripDependedParam.argi + 2] = 0;
                this.mov_r_c(tmpreg, this.origin.add(this.inpos + oper.size + ripDependedParam.offset));
                this[oper.splits.join('_')](...oper.args);
                this.inpos += oper.size;
                return;
            }
        }
        oper.code.apply(this, oper.args);
        this.inpos += oper.size;
    }
    end() {
        const tmpreg = this.getUnusing();
        const originend = this.origin.add(this.codesize);
        if (tmpreg != null)
            this.jmp64(originend, tmpreg);
        else
            this.jmp64_notemp(originend);
    }
}
class ProcHacker {
    constructor(map) {
        this.map = map;
    }
    append(nmap) {
        const map = this.map;
        for (const key in nmap) {
            map[key] = nmap[key];
        }
        return this;
    }
    /**
     * @param subject name of hooking
     * @param key target symbol
     * @param offset offset from target
     * @param ptr target pointer
     * @param originalCode old codes
     * @param ignoreArea pairs of offset, ignores partial bytes.
     */
    check(subject, key, offset, ptr, originalCode, ignoreArea) {
        const buffer = ptr.getBuffer(originalCode.length);
        const diff = util_1.memdiff(buffer, originalCode);
        if (!util_1.memdiff_contains(ignoreArea, diff)) {
            console.error(colors.red(`${subject}: ${key}+0x${offset.toString(16)}: code unmatch`));
            console.error(colors.red(`[${util_1.hex(buffer)}] != [${util_1.hex(originalCode)}]`));
            console.error(colors.red(`diff: ${JSON.stringify(diff)}`));
            console.error(colors.red(`${subject}: skip `));
            return false;
        }
        else {
            return true;
        }
    }
    /**
     * @param subject for printing on error
     * @param key target symbol name
     * @param offset offset from target
     * @param originalCode bytes comparing before hooking
     * @param ignoreArea pair offsets to ignore of originalCode
     */
    nopping(subject, key, offset, originalCode, ignoreArea) {
        const ptr = this.map[key].add(offset);
        if (!ptr) {
            console.error(colors.red(`${subject}: skip, ${key} symbol not found`));
            return;
        }
        const size = originalCode.length;
        const unlock = new unlocker_1.MemoryUnlocker(ptr, size);
        if (this.check(subject, key, offset, ptr, originalCode, ignoreArea)) {
            dll_1.dll.vcruntime140.memset(ptr, 0x90, size);
        }
        unlock.done();
    }
    /**
     * @param key target symbol name
     * @param to call address
     */
    hookingRaw(key, to) {
        const origin = this.map[key];
        if (!origin)
            throw Error(`${key} symbol not found`);
        const REQUIRE_SIZE = 12;
        const codes = disassembler_1.disasm.process(origin, REQUIRE_SIZE);
        const out = new AsmMover(origin, codes.size);
        for (const oper of codes.operations) {
            const basename = oper.splits[0];
            if (basename === 'ret' || basename === 'jmp' || basename === 'call') {
                throw Error(`Too small area to patch, require=${REQUIRE_SIZE}, actual=${out.inpos}`);
            }
            out.asmFromOrigin(oper);
        }
        out.end();
        const original = out.alloc();
        const unlock = new unlocker_1.MemoryUnlocker(origin, codes.size);
        hacktool_1.hacktool.jump(origin, to, assembler_1.Register.rax, codes.size);
        unlock.done();
        return original;
    }
    /**
     * @param key target symbol name
     * @param to call address
     */
    hookingRawWithCallOriginal(key, to, keepRegister, keepFloatRegister) {
        const origin = this.map[key];
        if (!origin)
            throw Error(`${key} symbol not found`);
        const REQUIRE_SIZE = 12;
        const codes = disassembler_1.disasm.process(origin, REQUIRE_SIZE);
        const out = new AsmMover(origin, codes.size);
        for (const reg of keepRegister) {
            out.freeregs.add(reg);
        }
        out.saveAndCall(to, keepRegister, keepFloatRegister);
        for (const oper of codes.operations) {
            const basename = oper.splits[0];
            if (basename === 'ret' || basename === 'jmp' || basename === 'call') {
                throw Error(`Too small area to patch, require=${REQUIRE_SIZE}, actual=${out.inpos}`);
            }
            out.asmFromOrigin(oper);
        }
        out.end();
        const unlock = new unlocker_1.MemoryUnlocker(origin, codes.size);
        hacktool_1.hacktool.jump(origin, out.alloc(), assembler_1.Register.rax, codes.size);
        unlock.done();
    }
    /**
     * @param key target symbol name
     * @param to call address
     */
    hooking(key, returnType, opts, ...params) {
        return callback => {
            const to = makefunc_1.makefunc.np(callback, returnType, opts, ...params);
            return makefunc_1.makefunc.js(this.hookingRaw(key, to), returnType, opts, ...params);
        };
    }
    /**
     * @param subject for printing on error
     * @param key target symbol name
     * @param offset offset from target
     * @param newCode call address
     * @param tempRegister using register to call
     * @param call true - call, false - jump
     * @param originalCode bytes comparing before hooking
     * @param ignoreArea pair offsets to ignore of originalCode
     */
    patching(subject, key, offset, newCode, tempRegister, call, originalCode, ignoreArea) {
        let ptr = this.map[key];
        if (!ptr) {
            console.error(colors.red(`${subject}: skip, ${key} symbol not found`));
            return;
        }
        ptr = ptr.add(offset);
        if (!ptr) {
            console.error(colors.red(`${subject}: skip`));
            return;
        }
        const size = originalCode.length;
        const unlock = new unlocker_1.MemoryUnlocker(ptr, size);
        if (this.check(subject, key, offset, ptr, originalCode, ignoreArea)) {
            hacktool_1.hacktool.patch(ptr, newCode, tempRegister, size, call);
        }
        unlock.done();
    }
    /**
     * @param subject for printing on error
     * @param key target symbol name
     * @param offset offset from target
     * @param jumpTo jump address
     * @param tempRegister using register to jump
     * @param originalCode bytes comparing before hooking
     * @param ignoreArea pair offsets to ignore of originalCode
     */
    jumping(subject, key, offset, jumpTo, tempRegister, originalCode, ignoreArea) {
        let ptr = this.map[key];
        if (!ptr) {
            console.error(colors.red(`${subject}: skip, ${key} symbol not found`));
            return;
        }
        ptr = ptr.add(offset);
        const size = originalCode.length;
        const unlock = new unlocker_1.MemoryUnlocker(ptr, size);
        if (this.check(subject, key, offset, ptr, originalCode, ignoreArea)) {
            hacktool_1.hacktool.jump(ptr, jumpTo, tempRegister, size);
        }
        unlock.done();
    }
    write(key, offset, asm) {
        const buffer = asm.buffer();
        const ptr = this.map[key].add(offset);
        const unlock = new unlocker_1.MemoryUnlocker(ptr, buffer.length);
        ptr.writeBuffer(buffer);
        unlock.done();
    }
    /**
     * make the native function as a JS function.
     *
     * wrapper codes are not deleted permanently.
     * do not use it dynamically.
     *
     * @param returnType RawTypeId or *Pointer
     * @param params RawTypeId or *Pointer
     */
    js(key, returnType, opts, ...params) {
        return makefunc_1.makefunc.js(this.map[key], returnType, opts, ...params);
    }
    /**
     * get symbols from cache.
     * if symbols don't exist in cache. it reads pdb.
     */
    static load(cacheFilePath, names) {
        return new ProcHacker(core_1.pdb.getList(cacheFilePath, {}, names));
    }
}
exports.ProcHacker = ProcHacker;
