// bdsx/sharedpointer.js
"use strict";
var SharedPtrBase_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SharedPointer = exports.SharedPtr = exports.SharedPtrBase = void 0;
const tslib_1 = require("tslib");
const capi_1 = require("./capi");
const common_1 = require("./common");
const core_1 = require("./core");
const makefunc_1 = require("./makefunc");
const nativeclass_1 = require("./nativeclass");
const nativetype_1 = require("./nativetype");
const singleton_1 = require("./singleton");
let SharedPtrBase = SharedPtrBase_1 = class SharedPtrBase extends nativeclass_1.NativeClass {
    [nativetype_1.NativeType.ctor]() {
        this.useRef = 1;
        this.weakRef = 1;
    }
    addRef() {
        this.interlockedIncrement32(8); // useRef
        this.interlockedIncrement32(16); // weakRef
    }
    release() {
        if (this.interlockedDecrement32(0x8) === 0) {
            this._Destroy();
        }
        if (this.interlockedDecrement32(0xc) === 0) {
            this._DeleteThis();
        }
    }
    _DeleteThis() {
        common_1.abstract();
    }
    _Destroy() {
        common_1.abstract();
    }
    static make(type) {
        return sharedPtrBaseSingleton.newInstance(type, () => {
            class SharedPtrBaseImpl extends SharedPtrBase_1 {
            }
            SharedPtrBaseImpl.define({ value: type });
            return SharedPtrBaseImpl;
        });
    }
};
tslib_1.__decorate([
    nativeclass_1.nativeField(core_1.VoidPointer)
], SharedPtrBase.prototype, "vftable", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint32_t)
], SharedPtrBase.prototype, "useRef", void 0);
tslib_1.__decorate([
    nativeclass_1.nativeField(nativetype_1.uint32_t)
], SharedPtrBase.prototype, "weakRef", void 0);
SharedPtrBase = SharedPtrBase_1 = tslib_1.__decorate([
    nativeclass_1.nativeClass()
], SharedPtrBase);
exports.SharedPtrBase = SharedPtrBase;
SharedPtrBase.prototype._Destroy = makefunc_1.makefunc.js([0], makefunc_1.RawTypeId.Void, { this: SharedPtrBase });
SharedPtrBase.prototype._DeleteThis = makefunc_1.makefunc.js([8], makefunc_1.RawTypeId.Void, { this: SharedPtrBase });
const sharedPtrBaseSingleton = new singleton_1.Singleton();
const sizeOfSharedPtrBase = SharedPtrBase[nativetype_1.NativeType.size];
/**
 * wrapper for std::shared_ptr
 */
class SharedPtr extends nativeclass_1.NativeClass {
    [nativetype_1.NativeType.ctor]() {
        this.p = null;
        this.ref = null;
    }
    [nativetype_1.NativeType.dtor]() {
        if (this.ref !== null)
            this.ref.release();
    }
    [nativetype_1.NativeType.ctor_copy](value) {
        this.p = value.p;
        this.ref = value.ref;
        if (this.ref !== null)
            this.ref.addRef();
    }
    [nativetype_1.NativeType.ctor_move](value) {
        this.p = value.p;
        this.ref = value.ref;
        value.p = null;
        value.ref = null;
    }
    ctor_move(value) {
        this.p = value.p;
        this.ref = value.ref;
        value.ref = null;
    }
    assign(value) {
        this[nativetype_1.NativeType.dtor]();
        this[nativetype_1.NativeType.ctor_copy](value);
        return this;
    }
    assign_move(value) {
        this[nativetype_1.NativeType.dtor]();
        this[nativetype_1.NativeType.ctor_move](value);
        return this;
    }
    exists() {
        return this.ref !== null;
    }
    addRef() {
        this.ref.addRef();
    }
    assignTo(dest) {
        const ctor = this.constructor;
        const ptr = dest.as(ctor);
        ptr.assign(this);
    }
    dispose() {
        if (this.ref !== null) {
            this.ref.release();
            this.ref = null;
        }
        this.p = null;
    }
    static make(cls) {
        const clazz = cls;
        return sharedPtrSingleton.newInstance(cls, () => {
            const Base = SharedPtrBase.make(clazz);
            class TypedSharedPtr extends SharedPtr {
                create(vftable) {
                    const size = Base[nativetype_1.NativeType.size];
                    if (size === null)
                        throw Error(`cannot allocate the non sized class`);
                    this.ref = capi_1.capi.malloc(size).as(Base);
                    this.ref.vftable = vftable;
                    this.ref.construct();
                    this.p = this.ref.addAs(clazz, sizeOfSharedPtrBase);
                }
            }
            TypedSharedPtr.define({
                p: clazz.ref(),
                ref: Base.ref(),
            });
            return TypedSharedPtr;
        });
    }
}
exports.SharedPtr = SharedPtr;
const sharedPtrSingleton = new singleton_1.Singleton();
/**
 * @deprecated
 */
class SharedPointer extends core_1.StaticPointer {
    constructor(sharedptr) {
        super(sharedptr.p);
        this.sharedptr = sharedptr;
    }
    assignTo(dest) {
        this.assignTo(dest);
    }
    dispose() {
        this.sharedptr.dispose();
    }
}
exports.SharedPointer = SharedPointer;
