// bdsx/tester.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tester = void 0;
const source_map_support_1 = require("./source-map-support");
const util_1 = require("./util");
const colors = require("colors");
let passed = 0;
let testnum = 1;
let testcount = 0;
class Tester {
    constructor() {
        this.subject = '';
        this.errored = false;
        this.done = false;
    }
    log(message) {
        console.log(`[test/${this.subject}] ${message}`);
    }
    _error(message, errorpos) {
        console.error(colors.red(`[test/${this.subject}] failed. ${message}`));
        console.error(colors.red(errorpos));
        if (!this.errored) {
            if (this.done) {
                passed--;
                console.error(colors.red(`[test] FAILED (${passed}/${testcount})`));
            }
            this.errored = true;
            Tester.errored = true;
        }
    }
    error(message, stackidx = 2) {
        const stack = Error().stack;
        this._error(message, source_map_support_1.remapStackLine(util_1.getLineAt(stack, stackidx)).stackLine);
    }
    processError(err) {
        const stack = (source_map_support_1.remapError(err).stack || '').split('\n');
        this._error(err.message, stack[1]);
        console.error(stack.slice(2).join('\n'));
    }
    fail() {
        this.error('', 3);
    }
    assert(cond, message) {
        if (!cond)
            this.error(message, 3);
    }
    static async test(tests) {
        await new Promise(resolve => setTimeout(resolve, 100)); // run after examples
        // pass one tick, wait until result of the list command example
        {
            const system = server.registerSystem(0, 0);
            await new Promise(resolve => {
                system.update = () => {
                    resolve();
                    system.update = undefined;
                };
            });
        }
        console.log(`[test] node: ${process.versions.node}`);
        console.log('[test] engine: ' + process.jsEngine + '@' + process.versions[process.jsEngine]);
        const testlist = Object.entries(tests);
        testcount += testlist.length;
        for (const [subject, test] of testlist) {
            const tester = new Tester;
            try {
                console.log(`[test] (${testnum++}/${testcount}) ${subject}`);
                tester.subject = subject;
                tester.errored = false;
                await test.call(tester);
                if (!tester.errored)
                    passed++;
                tester.done = true;
            }
            catch (err) {
                tester.processError(err);
            }
        }
        if (passed !== testcount) {
            console.error(colors.red(`[test] FAILED (${passed}/${testcount})`));
        }
        else {
            console.log(`[test] PASSED (${passed}/${testcount})`);
        }
    }
}
exports.Tester = Tester;
Tester.errored = false;
